{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('/home/orenmil/raid/mds/240801_standalone_blood_aging/code')\n",
    "# print(os.getcwd())\n",
    "\n",
    "\n",
    "import sys\n",
    "import shutil\n",
    "# caution: path[0] is reserved for script path (or '' in REPL)\n",
    "\n",
    "import importlib\n",
    "import itertools\n",
    "import metacells as mc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.cluster.hierarchy\n",
    "import scipy.spatial.distance\n",
    "import matplotlib\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "import pathlib\n",
    "import logging\n",
    "import re\n",
    "import collections\n",
    "import sklearn\n",
    "import pickle\n",
    "import sklearn.neighbors\n",
    "\n",
    "# from generic im\n",
    "# port gene_module_utils\n",
    "from generic import generic_utils\n",
    "from mds import mds_analysis_params\n",
    "# from mds import ipssm_utils\n",
    "from mds import mds_analysis\n",
    "from generic import mc_utils\n",
    "from generic import hg38_utils\n",
    "from mds import arch_mutation_interface_and_utils\n",
    "from mds import clinical_data_interface_and_utils\n",
    "from sc_rna_seq_preprocessing import sc_rna_seq_preprocessing_params\n",
    "from mds import lateral_and_noisy_genes\n",
    "\n",
    "plt.rcParams[\"patch.force_edgecolor\"] = False\n",
    "plt.rcParams['patch.linewidth'] = 0\n",
    "plt.rcParams['patch.edgecolor'] = 'none'\n",
    "# plt.rcParams['scatter.edgecolors'] = 'black' # didnt affect sb.scatterplot\n",
    "\n",
    "%matplotlib widget\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "def get_mds_params():\n",
    "    return mds_analysis_params.MDS_ANALYSIS_PARAMS\n",
    "\n",
    "\n",
    "def get_sc_rna_seq_preprocessing_params():\n",
    "    return sc_rna_seq_preprocessing_params.SC_RNA_SEQ_PREPROCESSING_PARAMS\n",
    "\n",
    "plt.subplots()\n",
    "plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n2_mc_ad = mc_utils.get_atlas_ad(get_mds_params()['nimrod_oren_atlas'])\n",
    "n2_c_ad = ad.read_h5ad(get_mds_params()['nimrod_oren_atlas']['c_ad_file_path'])\n",
    "\n",
    "# the following call is needed to change the unassigned from white to light grey...\n",
    "mc_utils.add_c_state_and_mc_c_state_stats(\n",
    "    n2_mc_ad, n2_c_ad, \n",
    "    cell_state_and_info_list=get_mds_params()['pb_cd34_enriched_cell_state_and_c_info_list'], \n",
    "    mask_and_info_list=get_mds_params()['pb_cd34_enriched_mask_and_c_info_list'], \n",
    "    cell_type_colors_csv_file_path=get_mds_params()['cell_type_colors_csv_file_path'],\n",
    "    only_add_c_state=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig_sup_4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "allow_skipping_plots = True\n",
    "allow_skipping_plots = False\n",
    "\n",
    "from mds import pb_cd34_c_score_threshs\n",
    "\n",
    "c_ad = n2_c_ad.copy()\n",
    "b_states = ['aged-B?', 'CCDC88A-high-B', 'naive-B', 'memory-B']\n",
    "monocyte_states = ['cMonocyte', 'intermMonocyte', 'ncMonocyte']\n",
    "dc_states = ['AS-DC', 'pDC', 'cDC?', 'high_cMonocyte_sig_cDC2-prog?', 'cDC2-prog?', 'DC-prog?', 'cDC1']\n",
    "nkt_states = ['NK', 'T']\n",
    "c_ad.obs['c_state'].replace({\n",
    "    **{x: 'B' for x in b_states},\n",
    "    **{x: 'Monocyte' for x in monocyte_states},\n",
    "    **{x: 'DC' for x in dc_states},\n",
    "    **{x: 'NKT' for x in nkt_states},\n",
    "    'monocyte_B_doublet': 'Doublet',\n",
    "    'plasmablast_ighm_ighg': 'plasmablast',\n",
    "    'state_unassigned': 'unassigned',\n",
    "}, inplace=True)\n",
    "\n",
    "\n",
    "c_ad.obs = generic_utils.merge_preserving_df1_index_and_row_order(c_ad.obs.drop(columns='c_state_color'), pd.read_csv(get_mds_params()['cell_type_colors_csv_file_path']).rename(columns={'cell_type': 'c_state', 'color': 'c_state_color'}))\n",
    "c_ad.obs['c_state'].value_counts()\n",
    "\n",
    "\n",
    "\n",
    "def get_name_to_mask_func(c_ad):\n",
    "    state_series = c_ad.obs['c_state'].astype(str)\n",
    "    return {\n",
    "        'HSPC': state_series.isin(mds_analysis_params.PB_HSPC_STATE_NAMES),\n",
    "        'non_lymphoid_HSPC': state_series.isin(mds_analysis_params.PB_NON_LYMPHOID_HSPC_STATE_NAMES),\n",
    "        **{x: state_series == x for x in state_series.unique()},\n",
    "    }\n",
    "\n",
    "scatter_name_to_info = {\n",
    "    'separate_HSPC_from_DC_and_monocyte_and_B': dict(\n",
    "        x='higher_in_dc_than_all_hspcs',\n",
    "        y='higher_in_b_than_all_other_hspcs',\n",
    "        x_thresh=pb_cd34_c_score_threshs.HIGHER_IN_DC_THAN_ALL_HSPCS_THRESH,\n",
    "        y_thresh=pb_cd34_c_score_threshs.HIGHER_IN_B_THAN_ALL_OTHER_HSPCS_THRESH,\n",
    "        background_name='all',\n",
    "        # gates_to_target_mask_name={('both_low',): 'HSPC'},\n",
    "        skip_plot=allow_skipping_plots,\n",
    "    ),\n",
    "    'separate_HSPC_from_NKT_and_endothel': dict(\n",
    "        x='nkt_somewhat_specific',\n",
    "        y='endothel_somewhat_specific',\n",
    "        x_thresh=pb_cd34_c_score_threshs.NKT_SOMEWHAT_SPECIFIC_THRESH,\n",
    "        y_thresh=pb_cd34_c_score_threshs.HIGHER_IN_ENDOTHEL_THAN_ALL_EXCEPT_UNKNOWN_THRESH,\n",
    "        background_name='non-B/Monocyte/DC',\n",
    "        get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: scatter_name_to_mask_dict['separate_HSPC_from_DC_and_monocyte_and_B']['both_low'],\n",
    "        gates_to_target_mask_name={('both_low',): 'HSPC'},\n",
    "        skip_plot=allow_skipping_plots,\n",
    "    ),\n",
    "    'identify_pro_B_and_pre_B': dict(\n",
    "        x='higher_in_pre_b_than_all_hspcs',\n",
    "        y='higher_in_pre_b_than_pro_b',\n",
    "        x_thresh=pb_cd34_c_score_threshs.HIGHER_IN_PRE_B_THAN_ALL_HSPCS_THRESH,\n",
    "        y_thresh=pb_cd34_c_score_threshs.HIGHER_IN_PRE_B_THAN_PRO_B_THRESH,\n",
    "        background_name='HSPC',\n",
    "        # get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: scatter_name_to_mask_dict['separate_HSPC_from_NKT_and_endothel']['both_low'],\n",
    "        get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: c_ad.obs['not_dc_nkt_monocyte_endothel_b_lamp3'],\n",
    "        gates_to_target_mask_name={('both_high',): 'pre-B?', ('high_x_low_y',): 'pro-B?'},\n",
    "        skip_plot=allow_skipping_plots,\n",
    "    ),\n",
    "    'separate_non_lymphoid_HSPC_from_CLP_and_NKTDP': dict(\n",
    "        x='higher_in_clp_m_than_all_myeloid_hspcs',\n",
    "        y='higher_in_nktdp_than_all_myeloid_hspcs',\n",
    "        x_thresh=pb_cd34_c_score_threshs.HIGHER_IN_CLP_M_THAN_ALL_MYELOID_HSPCS_THRESH,\n",
    "        y_thresh=pb_cd34_c_score_threshs.HIGHER_IN_NKTDP_THAN_ALL_MYELOID_HSPCS_THRESH,\n",
    "        background_name='non-B-prog HSPC',\n",
    "        get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: (\n",
    "            scatter_name_to_mask_dict['identify_pro_B_and_pre_B']['both_low'] |\n",
    "            scatter_name_to_mask_dict['identify_pro_B_and_pre_B']['low_x_high_y']\n",
    "        ),\n",
    "        gates_to_target_mask_name={('both_low',): 'non_lymphoid_HSPC'},\n",
    "        skip_plot=allow_skipping_plots,\n",
    "    ),\n",
    "    'indentify_NKTDP_and_CLP': dict(\n",
    "        x='higher_in_clp_m_than_all_myeloid_hspcs',\n",
    "        y='higher_in_nktdp_than_clp',\n",
    "        x_thresh=pb_cd34_c_score_threshs.HIGHER_IN_CLP_M_THAN_ALL_MYELOID_HSPCS_THRESH,\n",
    "        y_thresh=pb_cd34_c_score_threshs.HIGHER_IN_NKTDP_THAN_CLP_THRESH,\n",
    "        background_name='high-NKTDP-sig or high-CLP-sig non-B-prog HSPC',\n",
    "        get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: (\n",
    "            scatter_name_to_mask_dict['separate_non_lymphoid_HSPC_from_CLP_and_NKTDP']['low_x_high_y'] |\n",
    "            scatter_name_to_mask_dict['separate_non_lymphoid_HSPC_from_CLP_and_NKTDP']['both_high'] |\n",
    "            scatter_name_to_mask_dict['separate_non_lymphoid_HSPC_from_CLP_and_NKTDP']['high_x_low_y']\n",
    "        ),\n",
    "        gates_to_target_mask_name={('low_x_high_y','both_high'): 'NKTDP', ('high_x_low_y',): 'CLP'},\n",
    "        skip_plot=allow_skipping_plots,\n",
    "    ),\n",
    "    'identify_GMP-L_and_BEMP': dict(\n",
    "        x='higher_in_bemp_than_mebemp_l_and_eryp',\n",
    "        y='higher_in_gmp_l_than_all_other_hspcs',\n",
    "        x_thresh=pb_cd34_c_score_threshs.HIGHER_IN_BEMP_THAN_MEBEMP_L_AND_ERYP_THRESH,\n",
    "        y_thresh=pb_cd34_c_score_threshs.HIGHER_IN_GMP_L_THAN_ALL_OTHER_HSPCS_THRESH,\n",
    "        background_name='non-lymphoid HSPC',\n",
    "        # get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: c_ad.obs['not_dc_nkt_monocyte_endothel_b_lamp3'],\n",
    "        get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: scatter_name_to_mask_dict['separate_non_lymphoid_HSPC_from_CLP_and_NKTDP']['both_low'],\n",
    "        gates_to_target_mask_name={('high_x_low_y',): 'BEMP', ('low_x_high_y','both_high'): 'GMP-L'},\n",
    "        skip_plot=allow_skipping_plots,\n",
    "    ),\n",
    "    'identify_MKP_and_separate_by_ERYP_sig': dict(\n",
    "        x='higher_in_mkp_than_mebemp_l_and_eryp',\n",
    "        y='higher_in_ep_than_mpp',\n",
    "        x_thresh=pb_cd34_c_score_threshs.HIGHER_IN_MKP_THAN_MEBEMP_L_AND_ERYP_THRESH,\n",
    "        y_thresh=pb_cd34_c_score_threshs.HIGHER_IN_EP_THAN_MPP_THRESH,\n",
    "        background_name='non-lymphoid/BEMP/GMP-L HSPC',\n",
    "        # get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: c_ad.obs['not_dc_nkt_monocyte_endothel_b_lamp3'],\n",
    "        get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: scatter_name_to_mask_dict['identify_GMP-L_and_BEMP']['both_low'],\n",
    "        gates_to_target_mask_name={('high_x_low_y','both_high'): 'MKP'},\n",
    "        skip_plot=allow_skipping_plots,\n",
    "    ),\n",
    "    'identify_MEBEMP-L_and_HSC_MPP': dict(\n",
    "        x='higher_in_mebemp_l_than_mpp',\n",
    "        y='higher_in_ep_than_mpp',\n",
    "        x_thresh=pb_cd34_c_score_threshs.HIGHER_IN_MEBEMP_L_THAN_MPP_THRESH,\n",
    "        y_thresh=pb_cd34_c_score_threshs.HIGHER_IN_EP_THAN_MPP_THRESH,\n",
    "        background_name='non-lymphoid/BEMP/GMP-L/MKP HSPC',\n",
    "        # get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: c_ad.obs['not_dc_nkt_monocyte_endothel_b_lamp3'],\n",
    "        get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: (\n",
    "            scatter_name_to_mask_dict['identify_MKP_and_separate_by_ERYP_sig']['both_low'] |\n",
    "            scatter_name_to_mask_dict['identify_MKP_and_separate_by_ERYP_sig']['low_x_high_y']\n",
    "        ),\n",
    "        gates_to_target_mask_name={('high_x_low_y',): 'MEBEMP-L', ('both_low',): 'HSC_MPP'},\n",
    "        skip_plot=allow_skipping_plots,\n",
    "    ),\n",
    "    'identify_ERYP_and_high_MPP_sig_ERYP': dict(\n",
    "        x='higher_in_mpp_than_ep',\n",
    "        y='higher_in_mebemp_l_than_mpp',\n",
    "        x_thresh=pb_cd34_c_score_threshs.HIGHER_IN_MPP_THAN_EP_THRESH,\n",
    "        y_thresh=pb_cd34_c_score_threshs.HIGHER_IN_MEBEMP_L_THAN_MPP_THRESH,\n",
    "        background_name='high-ERYP-sig non-lymphoid/BEMP/GMP-L/MKP HSPC',\n",
    "        # get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: c_ad.obs['not_dc_nkt_monocyte_endothel_b_lamp3'],\n",
    "        get_background_mask_func=lambda scatter_name_to_mask_dict, c_ad: (\n",
    "            scatter_name_to_mask_dict['identify_MEBEMP-L_and_HSC_MPP']['both_high'] |\n",
    "            scatter_name_to_mask_dict['identify_MEBEMP-L_and_HSC_MPP']['low_x_high_y']\n",
    "        ),\n",
    "        gates_to_target_mask_name={('low_x_high_y','both_low'): 'ERYP', ('both_high','high_x_low_y'): 'high_MPP_sig_ERYP'},\n",
    "        # skip_plot=allow_skipping_plots,\n",
    "    ),\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "palette = mc_utils.get_palette(c_ad, 'c_state')\n",
    "assert 'white' not in palette.values()\n",
    "\n",
    "name_to_mask = get_name_to_mask_func(c_ad)\n",
    "\n",
    "mds_analysis.plot_gate_c_state_scatterplots(\n",
    "    c_ad=c_ad,\n",
    "    out_dir_path=out_dir_path,\n",
    "    scatter_name_to_info=scatter_name_to_info,\n",
    "    palette=palette,\n",
    "    name_to_mask=name_to_mask,\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig_sup_4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# ordered_c_states = [x for x in mds_analysis_params.ORDERED_PB_HSPC_STATE_NAMES if x in c_ad.obs['c_state'].unique()]\n",
    "ordered_hspc_c_states = [x for x in mds_analysis_params.ORDERED_PB_HSPC_STATE_NAMES if x in c_ad.obs['c_state'].unique()]\n",
    "ordered_non_hspc_c_states = [x for x in mds_analysis_params.ORDERED_PB_NON_HSPC_STATE_NAMES if x in c_ad.obs['c_state'].unique()]\n",
    "# raise\n",
    "state_color_df = pd.DataFrame({'state': ordered_hspc_c_states[::-1] + ordered_non_hspc_c_states})\n",
    "# raise\n",
    "state_color_df = generic_utils.merge_preserving_df1_index_and_row_order(state_color_df, pd.read_csv(get_mds_params()['cell_type_colors_csv_file_path']).rename(columns={'cell_type': 'state'}))\n",
    "fig, ax = plt.subplots(figsize=(10,10))\n",
    "for x in state_color_df.to_records(index=False).tolist():\n",
    "    ax.scatter([0.5], [0.5], s=40, color=x[1], label=x[0])\n",
    "\n",
    "ax.legend(fontsize='large')\n",
    "sb.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1), fontsize='large')\n",
    "fig.tight_layout()\n",
    "fig.savefig(\n",
    "    os.path.join(out_dir_path, f'EDF_9A_cell_state_circle_legend.png'), \n",
    "    dpi=300,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close('all')\n",
    "out_dir_path = 'temp/_fig_sup_4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "add_ax_titles = True\n",
    "add_ax_titles = False\n",
    "\n",
    "\n",
    "mc_ad = n2_mc_ad.copy()\n",
    "b_states = ['aged-B?', 'CCDC88A-high-B', 'naive-B', 'memory-B']\n",
    "monocyte_states = ['cMonocyte', 'intermMonocyte', 'ncMonocyte']\n",
    "dc_states = ['AS-DC', 'pDC', 'cDC?']\n",
    "nkt_states = ['NK', 'T']\n",
    "mc_ad.obs['state'].replace({\n",
    "    **{x: 'B' for x in b_states},\n",
    "    **{x: 'Monocyte' for x in monocyte_states},\n",
    "    **{x: 'DC' for x in dc_states},\n",
    "    **{x: 'NKT' for x in nkt_states},\n",
    "    'CLP_NKTDP_intermediate?': 'Doublet', # unsure, but anyway only 3 of those currently...\n",
    "}, inplace=True)\n",
    "mc_ad.obs = generic_utils.merge_preserving_df1_index_and_row_order(mc_ad.obs.drop(columns='state_color'), pd.read_csv(get_mds_params()['cell_type_colors_csv_file_path']).rename(columns={'cell_type': 'state', 'color': 'state_color'}))\n",
    "\n",
    "c_state_count_series = n2_c_ad.obs.loc[n2_c_ad.obs['c_state'].isin(mds_analysis_params.PB_HSPC_STATE_NAMES), 'c_state'].astype(str).value_counts(normalize=True)\n",
    "common_c_states = list((c_state_count_series > 0.015).loc[lambda x: x].index)\n",
    "fig, ax = mc_utils.plot_manifold_umap(mc_ad, legend=True, move_legend=True, legend_labels_ordered=mds_analysis_params.ORDERED_CELL_STATE_NAMES)\n",
    "generic_utils.make_all_spines_and_x_and_y_axes_invisible(ax)\n",
    "# ax.get_yaxis().set_visible(False)\n",
    "# ax.get_xaxis().set_visible(False)\n",
    "ordered_common_c_states = [\n",
    "    'HSC_MPP',\n",
    "    'MEBEMP-L',\n",
    "    'ERYP',\n",
    "    'BEMP',\n",
    "    'CLP',\n",
    "    'NKTDP',\n",
    "]\n",
    "assert set(ordered_common_c_states) == set(common_c_states)\n",
    "fig.savefig(os.path.join(out_dir_path, 'EDF_9B_cHSPC_79_normal_atlas_umap.png'))\n",
    "fig, axes = plt.subplots(nrows=3, ncols=2, figsize=(4, 7), gridspec_kw=dict(hspace=0.3))\n",
    "fig.subplots_adjust(top=1.1, bottom=0.01, left=0.01, right=0.99) # seems like top is ignored.\n",
    "for c_state, ax in zip(ordered_common_c_states, axes.flatten()):\n",
    "    plot_manifold_umap_kwargs = dict(\n",
    "        mc_ad=mc_ad, color_by=f'c_state_{c_state}_frac', get_palette_kwargs=dict(numeric_colormap_name='gist_heat_r'), \n",
    "        # plot_colorbar=True, \n",
    "        hue_norm=(0,1),\n",
    "        fig_colorbar_kwargs=dict(label=f'{c_state} fraction'),\n",
    "        s=6,\n",
    "    )\n",
    "    fig, ax = mc_utils.plot_manifold_umap(ax=ax, **plot_manifold_umap_kwargs)\n",
    "    generic_utils.make_all_spines_and_x_and_y_axes_invisible(ax)\n",
    "    fig.subplots_adjust(top=0.9)\n",
    "    if add_ax_titles:\n",
    "        ax.set_title(c_state)\n",
    "ax_title_repr = '' if add_ax_titles else '_without_titles'\n",
    "fig.savefig(os.path.join(out_dir_path, f'EDF_9B_cHSPC_79_normal_atlas_umap_colored_by_c_state_frac{ax_title_repr}.png'), dpi=300)\n",
    "\n",
    "fig, ax = mc_utils.plot_manifold_umap(plot_colorbar=True, **plot_manifold_umap_kwargs)\n",
    "fig.savefig(os.path.join(out_dir_path, f'EDF_9B_cHSPC_79_normal_atlas_umap_colored_by_c_state_with_colorbar.png'), dpi=300)\n",
    "plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig_sup_4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "all_cna_attrs_df_csv_file_path = get_mds_params()['all_cna_attrs_df_csv_file_path']\n",
    "all_cna_attrs_df = pd.read_csv(all_cna_attrs_df_csv_file_path)\n",
    "all_cna_attrs_df = all_cna_attrs_df[all_cna_attrs_df['chrom'].notna()]\n",
    "all_cna_attrs_df = all_cna_attrs_df[['donor_id', 'chrom', 'is_del', 'is_dup']].drop_duplicates()\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "all_cna_attrs_df = all_cna_attrs_df[all_cna_attrs_df['donor_id'].isin(all_ext_feature_df['donor_id'])]\n",
    "# all_ext_feature_df[all_ext_feature_df['cna_count'] > 0]\n",
    "# all_ext_feature_df.sort_values('cna_count', ascending=False).drop_duplicates(subset='donor_id')\n",
    "\n",
    "ordered_chrom_names = get_mds_params()['karyotype_estimation']['ordered_chrom_names_for_karyotype_estimation']\n",
    "\n",
    "flat_dicts = []\n",
    "for donor_id, curr_df in all_cna_attrs_df.groupby('donor_id'):\n",
    "    curr_dict = {x: 0 for x in ordered_chrom_names}\n",
    "    for _, row in curr_df.iterrows():\n",
    "        assert row[['is_dup', 'is_del']].any()\n",
    "        curr_dict[row['chrom']] += 0.5 if row['is_del'] else 1\n",
    "    flat_dicts.append({\n",
    "        'donor_id': donor_id,\n",
    "        **curr_dict,\n",
    "    })\n",
    "df = pd.DataFrame(flat_dicts)\n",
    "df.set_index('donor_id', inplace=True)\n",
    "df = df[ordered_chrom_names]\n",
    "assert df.isin([0,0.5,1]).all(axis=None), 'sorry, handling both dup and del in same chrom is not implemented'\n",
    "ordered_donor_ids = list((df > 0).sum(axis=1).sort_values().index)\n",
    "orig_donor_count = len(df)\n",
    "df = df.loc[ordered_donor_ids]\n",
    "assert len(df) == orig_donor_count\n",
    "\n",
    "# val_to_color = {0: 'white', 1: 'blue', 2: 'red'}\n",
    "# for col in df.columns:\n",
    "#     df[col] = df[col].map(val_to_color)\n",
    "\n",
    "# color_arr = generic_utils.get_color_arr(\n",
    "#     df, col_to_val_to_color={x: val_to_color for x in ordered_chrom_names},\n",
    "# )\n",
    "\n",
    "plt.close('all')\n",
    "fig, ax = plt.subplots(figsize=(2.7,5))\n",
    "# color_arr = df.values\n",
    "# color_df = pd.DataFrame()\n",
    "# ax.imshow(color_arr, aspect='auto')\n",
    "ugly_hack_cna_cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\n",
    "    '', [(0, 'white'), (0.5, 'blue'), (1, 'red')])\n",
    "df = df.T\n",
    "heatmap_obj = sb.heatmap(\n",
    "    df, ax=ax, cmap=ugly_hack_cna_cmap, square=True, cbar=False, \n",
    "    xticklabels=True, \n",
    "    # yticklabels=True,\n",
    ")\n",
    "fig.subplots_adjust(top=0.95, bottom=0.1)\n",
    "for _, spine in heatmap_obj.spines.items():\n",
    "    spine.set_visible(True)\n",
    "for x in range(len(df.index)):\n",
    "    ax.axhline(x, color='grey', alpha=0.7, linewidth=0.5)\n",
    "for x in range(len(df.columns)):\n",
    "    ax.axvline(x, color='grey', alpha=0.7, linewidth=0.5)\n",
    "ax.set_ylabel(None)\n",
    "ax.set_xlabel(None)\n",
    "if 0:\n",
    "    ax.legend(\n",
    "        handles=[matplotlib.patches.Patch(color=x[1], label=x[0]) for x in [('deletion', 'blue'), ('duplication', 'red')]],\n",
    "        # fontsize='large',\n",
    "    )\n",
    "    sb.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1), fontsize='large')\n",
    "# fig.tight_layout()\n",
    "fig.savefig(\n",
    "    os.path.join(out_dir_path, f'EDF_9C_cna_heatmap.png'), \n",
    "    dpi=300,\n",
    ")\n",
    "\n",
    "total_cna_count = (df > 0).sum().sum()\n",
    "print(f'total_cna_count: {total_cna_count}')\n",
    "assert total_cna_count == 38 # 240729\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig_sup_4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "final_feature_names_and_types_df = pd.read_csv(get_mds_params()['final_feature_names_and_types_df_csv_file_path'])\n",
    "\n",
    "list_of_max_pval_and_star_count = [\n",
    "    (0.001, 3),\n",
    "    (0.01, 2),\n",
    "    (0.05, 1),\n",
    "]\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "curr_feature_df = all_ext_feature_df[all_ext_feature_df['in_mds_cyto_groups_analysis']].copy()\n",
    "curr_feature_df['diag_class_and_sex'] = curr_feature_df['diagnosis_class'] + ' ' + curr_feature_df['donor_sex']\n",
    "\n",
    "final_feature_name_to_type = generic_utils.get_dict_mapping_one_df_column_to_other(final_feature_names_and_types_df, 'name', 'type')\n",
    "diag_class_and_sex_to_ref_diag_class_and_sex = {}\n",
    "ref_class = 'normal'\n",
    "for clas in ['cytopenia', 'MDS', 'MDS/MPN']:\n",
    "    for sex in ('male', 'female'):\n",
    "        clas_and_sex = f'{clas} {sex}'\n",
    "        if clas != ref_class:\n",
    "            diag_class_and_sex_to_ref_diag_class_and_sex[clas_and_sex] = f'{ref_class} {sex}'\n",
    "\n",
    "ordered_class_and_sex = [\n",
    "    'normal male',\n",
    "    'normal female',\n",
    "    'cytopenia male',\n",
    "    'cytopenia female',\n",
    "    'MDS male',\n",
    "    'MDS female',\n",
    "    'MDS/MPN male',\n",
    "    'MDS/MPN female',\n",
    "]\n",
    "class_and_sex_to_pos_in_stripplot = {\n",
    "    'normal male': 0,\n",
    "    'normal female': 1,\n",
    "    'cytopenia male': 2.5,\n",
    "    'cytopenia female': 3.5,\n",
    "    'MDS male': 5,\n",
    "    'MDS female': 6,\n",
    "    'MDS/MPN male': 7.5,\n",
    "    'MDS/MPN female': 8.5,\n",
    "}\n",
    "male_female_vline_xs = [0.5, 3, 5.5, 8]\n",
    "class_vline_xs = [1.75, 4.25, 6.75]\n",
    "assert pd.Series([class_and_sex_to_pos_in_stripplot[x] for x in ordered_class_and_sex]).is_monotonic_increasing\n",
    "\n",
    "flat_dicts = []\n",
    "for feature in sorted(final_feature_names_and_types_df.loc[final_feature_names_and_types_df['type'] == 'log_c_state_freq', 'name']):\n",
    "    # print(feature)\n",
    "    curr_df = curr_feature_df[curr_feature_df[feature].notna()]\n",
    "        \n",
    "        \n",
    "    for clas_and_sex, ref_clas_and_sex in diag_class_and_sex_to_ref_diag_class_and_sex.items():\n",
    "        clas_vals = curr_df.loc[curr_df['diag_class_and_sex'] == clas_and_sex, feature]\n",
    "        ref_vals = curr_df.loc[curr_df['diag_class_and_sex'] == ref_clas_and_sex, feature]\n",
    "        if (len(clas_vals) >= 1) and (len(ref_vals) >= 1):\n",
    "            mw_test_res = generic_utils.perform_mw_test(\n",
    "                clas_vals,\n",
    "                ref_vals,\n",
    "                alternative='two-sided',\n",
    "            )\n",
    "            flat_dicts.append({'diag_class_and_sex': clas_and_sex, 'feature': feature, 'feature_type': final_feature_name_to_type[feature], 'test_type': 'mw', 'pval': mw_test_res['pvalue']})\n",
    "\n",
    "diff_c_freq_across_diag_classes_df = pd.DataFrame(flat_dicts)\n",
    "diff_c_freq_across_diag_classes_df.sort_values('pval', inplace=True)\n",
    "mw_pval_mask = diff_c_freq_across_diag_classes_df['test_type'] == 'mw'\n",
    "diff_c_freq_across_diag_classes_df.loc[mw_pval_mask, 'mw_bh_adjusted_pval'] = scipy.stats.false_discovery_control(\n",
    "    diff_c_freq_across_diag_classes_df.loc[mw_pval_mask, 'pval'])\n",
    "\n",
    "curr_feature_df['diag_class_and_sex_pos_in_stripplot'] = curr_feature_df['diag_class_and_sex'].map(class_and_sex_to_pos_in_stripplot)\n",
    "np.random.seed(0)\n",
    "curr_feature_df['jittered_diag_class_and_sex_pos_in_stripplot'] = curr_feature_df['diag_class_and_sex_pos_in_stripplot'] + np.random.uniform(-0.3, 0.3, len(curr_feature_df))\n",
    "\n",
    "\n",
    "for attr in [\n",
    "    'log_c_CLP',\n",
    "    # 'log_c_pre-B?',\n",
    "    # 'log_c_pro-B?',\n",
    "    # 'log_c_MKP',\n",
    "    # 'log_c_GMP-L',\n",
    "]:\n",
    "    plt.close('all')\n",
    "    fig, ax = plt.subplots(figsize=(4, 2.2))\n",
    "    \n",
    "    assert curr_feature_df[attr].notna().all()\n",
    "\n",
    "    curr_observed_class_and_sex = set(curr_feature_df['diag_class_and_sex'].tolist())\n",
    "    curr_ordered_class_and_sex = [x for x in ordered_class_and_sex if x in curr_observed_class_and_sex]\n",
    "    assert curr_ordered_class_and_sex == ordered_class_and_sex\n",
    "\n",
    "    assert curr_feature_df['donor_id'].is_unique\n",
    "    print(f'len(curr_feature_df): {len(curr_feature_df)}')\n",
    "    \n",
    "    sb.scatterplot(\n",
    "        data=curr_feature_df, \n",
    "        x='jittered_diag_class_and_sex_pos_in_stripplot', \n",
    "        y=attr, \n",
    "        ax=ax,\n",
    "        hue='diagnosis_class', \n",
    "        palette=clinical_data_interface_and_utils.DIAGNOSIS_CLASS_TO_COLOR,\n",
    "        edgecolor='k',\n",
    "        s=20,\n",
    "    )\n",
    "    ax.get_xaxis().set_visible(False)\n",
    "\n",
    "    orig_y_lim = ax.get_ylim()\n",
    "    y_lim_range = orig_y_lim[1] - orig_y_lim[0]\n",
    "    ax.set_ylim(orig_y_lim[0], orig_y_lim[1] + 0.07 * y_lim_range)\n",
    "\n",
    "    for clas_and_sex in curr_ordered_class_and_sex:\n",
    "        if clas_and_sex in diag_class_and_sex_to_ref_diag_class_and_sex:\n",
    "            mw_test_res = generic_utils.perform_mw_test(\n",
    "                curr_feature_df.loc[curr_feature_df['diag_class_and_sex'] == clas_and_sex, attr],\n",
    "                curr_feature_df.loc[curr_feature_df['diag_class_and_sex'] == diag_class_and_sex_to_ref_diag_class_and_sex[clas_and_sex], attr],\n",
    "                alternative='two-sided',\n",
    "            )\n",
    "            curr_pval = mw_test_res['pvalue']\n",
    "\n",
    "            row_in_diff_feature_df = diff_c_freq_across_diag_classes_df[\n",
    "                (diff_c_freq_across_diag_classes_df['test_type'] == 'mw') &\n",
    "                (diff_c_freq_across_diag_classes_df['feature'] == attr) & \n",
    "                (diff_c_freq_across_diag_classes_df['diag_class_and_sex'] == clas_and_sex) \n",
    "            ]\n",
    "            assert len(row_in_diff_feature_df) == 1\n",
    "            row_in_diff_feature_df = row_in_diff_feature_df.iloc[0]\n",
    "            assert np.isclose(row_in_diff_feature_df['pval'], curr_pval)\n",
    "            assert np.isclose(row_in_diff_feature_df['pval'], curr_pval)\n",
    "            bh_adjusted_pval = row_in_diff_feature_df['mw_bh_adjusted_pval']\n",
    "\n",
    "            # raise\n",
    "            # assert  == final_feature_name_to_type[attr]\n",
    "\n",
    "            for max_pval, star_count in list_of_max_pval_and_star_count:\n",
    "                if bh_adjusted_pval < max_pval:\n",
    "                    x_pos = class_and_sex_to_pos_in_stripplot[clas_and_sex]\n",
    "                    x_pos = x_pos + 0.125 * (1 if clas_and_sex.endswith('female') else -1) \n",
    "                    ax.annotate(\n",
    "                        # '$' + '*' * star_count + '$',\n",
    "                        # '$*$' * star_count,\n",
    "                        '*' * star_count,\n",
    "                        # '$' + '\\\\star' * star_count + '$', # looks bad for some reason\n",
    "                        xy=(x_pos, 0.944), xycoords=('data', 'axes fraction'), color='k', ha='center', va='center', fontsize='large',\n",
    "                    )\n",
    "                    \n",
    "                    break\n",
    "\n",
    "\n",
    "\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_title(attr)\n",
    "    if 0:\n",
    "        ax.set_xticklabels(\n",
    "            ax.get_xticklabels(),\n",
    "            rotation=90, ha='center', \n",
    "            # rotation_mode='anchor', \n",
    "            # fontsize='small',\n",
    "        )\n",
    "    for x in male_female_vline_xs:\n",
    "        ax.axvline(x, color='grey', alpha=0.5, linestyle='--')\n",
    "    for x in class_vline_xs:\n",
    "        ax.axvline(x, color='black', alpha=0.9, linestyle='-')\n",
    "\n",
    "    # if plot_legend:\n",
    "    #     legend = ax.legend(fontsize='x-large')\n",
    "    #     legend.get_frame().set_alpha(None)\n",
    "    ax.get_legend().remove()\n",
    "    # fig.tight_layout()\n",
    "    fig.savefig(\n",
    "        os.path.join(out_dir_path, f'EDF_10B_{attr}_stripplot.png'), \n",
    "        dpi=300,\n",
    "    )\n",
    "    # break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig_sup_4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "show_donor_ids = True\n",
    "show_donor_ids = False\n",
    "\n",
    "plot_y_equals_x = True\n",
    "plot_y_equals_x = False\n",
    "\n",
    "curr_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "\n",
    "curr_df = curr_df[curr_df['in_within_state_sig_bio_rep_analysis']]\n",
    "print(f'len(curr_df): {len(curr_df)}')\n",
    "print(curr_df.drop_duplicates(subset='donor_id')['latest_diagnosis_class'].value_counts())\n",
    "\n",
    "final_feature_names_and_types_df = pd.read_csv(get_mds_params()['final_feature_names_and_types_df_csv_file_path'])\n",
    "\n",
    "bio_rep_df = sc_rna_seq_preprocessing_params.get_earliest_and_latest_bio_rep_df(curr_df)\n",
    "print(f'len(bio_rep_df): {len(bio_rep_df)}')\n",
    "\n",
    "flat_dicts = []\n",
    "for col in sorted([\n",
    "    # *final_feature_names_and_types_df.loc[final_feature_names_and_types_df['type'].isin(['gene_prog_excluding_mhc_ii', 'mhc_ii_gene_prog', 'log_c_state_freq']), 'name'],\n",
    "    'mebemp_l_hla_sig_MEBEMP-L_0.5',\n",
    "    'bemp_cd74_sig_BEMP_0.5',\n",
    "    's_phase_sig_MEBEMP-L_0.5',\n",
    "    \n",
    "    # 'prss2_sig_CLP_0.5',\n",
    "    # 'gmp_l_elane_sig_GMP-L_0.5',\n",
    "]):\n",
    "\n",
    "    x_col = f'{col}_early'\n",
    "    y_col = f'{col}_late'\n",
    "\n",
    "\n",
    "    curr_df = bio_rep_df[bio_rep_df[[x_col, y_col]].notna().all(axis=1)]\n",
    "    bio_rep_count = len(curr_df)\n",
    "    print(len(bio_rep_df))\n",
    "    print(f'{col} final bio_rep_count: {bio_rep_count}')\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(3.5,3.5))\n",
    "    sb.scatterplot(\n",
    "        data=curr_df,\n",
    "        x=x_col,\n",
    "        y=y_col,\n",
    "        hue='diagnosis_class_late',\n",
    "        palette=clinical_data_interface_and_utils.DIAGNOSIS_CLASS_TO_COLOR,\n",
    "        ax=ax,\n",
    "        legend=False,\n",
    "    )\n",
    "    slope, intercept, r_value, p_value, std_err_of_estimated_gradient = scipy.stats.linregress(\n",
    "        curr_df[x_col], curr_df[y_col],\n",
    "        # alternative='two-sided',\n",
    "    )\n",
    "    generic_utils.plot_line_on_ax(ax, slope=slope, offset=intercept, color='grey', alpha=0.5, linestyle='--')\n",
    "    if plot_y_equals_x:\n",
    "        generic_utils.plot_y_equals_x_line_on_ax(ax, color='k', alpha=0.5, linestyle=':')\n",
    "    linear_fit_repr = f'n={len(curr_df)}, r={r_value:.2f}, pval={p_value:.0E}'\n",
    "    print(f'slope={slope:.3f}, intercept={intercept:.3f}')\n",
    "    ax.set_title(linear_fit_repr)\n",
    "    if show_donor_ids:\n",
    "        for i, row in curr_df.iterrows():\n",
    "            ax.text(row[x_col], row[y_col], row['donor_id'])\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(os.path.join(out_dir_path, f'EDF_10D_{generic_utils.strip_special_chars_on_edges_and_replace_others_with_underscores(col)}_bio_reps.png'), dpi=300)\n",
    "    plt.close('all')\n",
    "    flat_dicts.append({\n",
    "        'feature': col,\n",
    "        'r_value': r_value,\n",
    "        'p_value': p_value,\n",
    "        'bio_rep_count': bio_rep_count,\n",
    "    })\n",
    "feature_bio_rep_df = pd.DataFrame(flat_dicts)\n",
    "feature_bio_rep_df.sort_values('p_value', inplace=True)\n",
    "#\n",
    "\n",
    "if 1:\n",
    "    print(list(feature_bio_rep_df.loc[feature_bio_rep_df['r_value'] < 0.5, 'feature']))\n",
    "    print(feature_bio_rep_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig_sup_4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "from mds import pb_cd34_c_score_threshs\n",
    "c_ad = n2_c_ad\n",
    "# c_ad = ad.read_h5ad(mds_analysis_params.get_mc_model_paths('final_mds_cyto_normal_excluding_atlas', dataset_name='illu_mds',\n",
    "# )['cells_with_metacell_attrs_ad'])\n",
    "mc_utils.add_c_state_and_mc_c_state_stats(\n",
    "    None, c_ad, \n",
    "    cell_state_and_info_list=get_mds_params()['pb_cd34_enriched_cell_state_and_c_info_list'], \n",
    "    mask_and_info_list=get_mds_params()['pb_cd34_enriched_mask_and_c_info_list'], \n",
    "    cell_type_colors_csv_file_path=get_mds_params()['cell_type_colors_csv_file_path'],\n",
    "    only_add_c_state=True,\n",
    ")\n",
    "\n",
    "c_mask = None\n",
    "\n",
    "palette = mc_utils.get_palette(c_ad, 'c_state')\n",
    "\n",
    "if c_mask is None:\n",
    "    c_mask = np.full(c_ad.n_obs, True)\n",
    "\n",
    "# curr_df = c_ad.obs[generic_utils.get_equal_contrib_mask([c_ad.obs['c_state'] == x for x in ['CLP', 'HSC_MPP']])].copy()\n",
    "curr_df = c_ad.obs[c_mask & c_ad.obs['c_state'].isin(['CLP', 'HSC_MPP'])].copy()\n",
    "curr_df['c_state'] = curr_df['c_state'].astype(str)\n",
    "\n",
    "plt.close('all')\n",
    "fig, ax = plt.subplots(figsize=(4,4))\n",
    "sb.histplot(\n",
    "    curr_df, \n",
    "    x='higher_in_clp_m_than_all_myeloid_hspcs', \n",
    "    # hue='c_state',\n",
    "    # palette=palette,\n",
    "    # common_norm=False,\n",
    "    # stat='proportion',\n",
    "    bins=100,\n",
    "    legend=False,\n",
    "    ax=ax,\n",
    ")\n",
    "for x in (pb_cd34_c_score_threshs.HIGHER_IN_CLP_M_THAN_ALL_MYELOID_HSPCS_CLP_E_LOW_THRESH, pb_cd34_c_score_threshs.HIGHER_IN_CLP_M_THAN_ALL_MYELOID_HSPCS_CLP_E_HIGH_THRESH):\n",
    "    ax.axvline(x=x, color='red', alpha=0.3)\n",
    "ax.set_title('CLP-sig across HSC_MPP and CLP')\n",
    "fig.tight_layout()\n",
    "fig.savefig(os.path.join(out_dir_path, 'EDF_11D_clp_e_definition_hist.png'), dpi=300)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "show_donor_ids = False\n",
    "show_donor_ids = True\n",
    "\n",
    "skip_plotting = True\n",
    "skip_plotting = False\n",
    "\n",
    "only_mpp_nd_ids_without_suffix = ['N405', 'N421']\n",
    "high_gmp_nd_ids_without_suffix = ['G12', 'N235_2', 'N367', 'N281_1', 'N250', 'N413', 'N403']\n",
    "high_bemp_nd_ids_without_suffix = ['N151', 'N192_1']\n",
    "high_clp_nd_ids_without_suffix = ['N387', 'N204_2']\n",
    "high_clp_low_mebemp_nd_ids_without_suffix = ['N203', 'N241', 'N283']\n",
    "comp_class_to_nd_ids_without_suffix = {\n",
    "    'only HSC/MPP': only_mpp_nd_ids_without_suffix,\n",
    "    'high GMP': high_gmp_nd_ids_without_suffix,\n",
    "    'high BEMP': high_bemp_nd_ids_without_suffix,\n",
    "    'high CLP': high_clp_nd_ids_without_suffix,\n",
    "    'high CLP low MEBEMP': high_clp_low_mebemp_nd_ids_without_suffix,\n",
    "}\n",
    "\n",
    "linkage_metric = 'correlation'\n",
    "linkage_metric = 'euclidean'\n",
    "\n",
    "linkage_method = 'average'\n",
    "linkage_method = 'ward' # scipy says \"Method 'ward' requires the distance metric to be Euclidean\"\n",
    "\n",
    "epsilon_for_cell_freq = get_mds_params()['epsilon_for_donor_log_c_state_freq']\n",
    "\n",
    "dont_show_normal = False\n",
    "dont_show_normal = True\n",
    "\n",
    "dont_show_normal_atlas = True\n",
    "dont_show_normal_atlas = False\n",
    "\n",
    "assert not (dont_show_normal_atlas and dont_show_normal)\n",
    "\n",
    "cluster_log_freqs = False\n",
    "cluster_log_freqs = True\n",
    "\n",
    "cluster_log_enrich = True\n",
    "cluster_log_enrich = False\n",
    "\n",
    "nn_dist_high_quantile = 0.98\n",
    "\n",
    "hclust_clust_count = None\n",
    "hclust_clust_count = 5\n",
    "\n",
    "\n",
    "order_by_k_means_k = 5\n",
    "order_by_k_means_k = None\n",
    "\n",
    "lower_ax_cols = [\n",
    "    'diagnosis_c',\n",
    "    'DNMT3A_max_mean_VAF',\n",
    "    'TET2_max_mean_VAF',\n",
    "    'SF3B1_max_mean_VAF',\n",
    "    'other max VAF',\n",
    "    'any_CNA',\n",
    "]\n",
    "\n",
    "order_donors_by_id = True\n",
    "order_donors_by_id = False\n",
    "\n",
    "order_donors_by_bleeding_date_and_exp_name = True\n",
    "order_donors_by_bleeding_date_and_exp_name = False\n",
    "\n",
    "discard_numbered_donor_id_tech_rep_suffix = False\n",
    "discard_numbered_donor_id_tech_rep_suffix = True\n",
    "\n",
    "verify_states_in_mds_analysis_params = False\n",
    "verify_states_in_mds_analysis_params = True\n",
    "\n",
    "all_ordered_state_names = [f'c_{x}' for x in mds_analysis_params.ORDERED_CELL_STATE_NAMES]\n",
    "ordered_pb_hspc_state_names = [f'c_{x}' for x in mds_analysis_params.ORDERED_PB_HSPC_STATE_NAMES]\n",
    "\n",
    "mutation_df = arch_mutation_interface_and_utils.get_minimal_all_mutation_df()\n",
    "mut_genes = sorted(mutation_df['gene'].drop_duplicates())\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "# all_ext_feature_df = pd.concat([pd.read_csv(x).assign(ext_donor_feature_df_file_path=x) for x in extended_df_csv_file_paths], ignore_index=True)\n",
    "# all_ext_feature_df = sc_rna_seq_preprocessing_params.get_df_with_numbered_donor_id(all_ext_feature_df)\n",
    "curr_feature_df = all_ext_feature_df.copy()\n",
    "\n",
    "curr_feature_df['atlas'] = curr_feature_df['dataset_name'] == 'ref'\n",
    "\n",
    "curr_feature_df['donor_age'] = curr_feature_df['donor_age'].astype(float)\n",
    "curr_feature_df['diagnosis_class'] = curr_feature_df['diagnosis_class'].astype(str)\n",
    "\n",
    "curr_feature_df['any_large_dels_or_dups'] = (curr_feature_df['cna_count'] > 0).astype(int)\n",
    "\n",
    "top_3_mut_genes = ['DNMT3A', 'TET2', 'SF3B1']\n",
    "mut_genes_except_top_3 = sorted(set(mut_genes) - set(top_3_mut_genes))\n",
    "curr_feature_df['max_mean_VAF_excluding_top_3'] = curr_feature_df[[f'{x}_max_mean_VAF' for x in mut_genes_except_top_3]].max(axis=1)\n",
    "\n",
    "observed_hspc_state_names = sorted(set(curr_feature_df.columns) & set(ordered_pb_hspc_state_names))\n",
    "log_hspc_freq_cols = [f'log_{x}' for x in observed_hspc_state_names]\n",
    "curr_feature_df[log_hspc_freq_cols] = np.log2(curr_feature_df[observed_hspc_state_names] + epsilon_for_cell_freq)\n",
    "\n",
    "unexpected_diagnosis_mask = ~curr_feature_df['diagnosis_class'].isin(['MDS', 'MDS/MPN', 'cytopenia', 'normal'])\n",
    "assert not unexpected_diagnosis_mask.any()\n",
    "\n",
    "curr_feature_df = curr_feature_df[curr_feature_df['in_mds_cyto_groups_analysis']]\n",
    "assert curr_feature_df['donor_id'].is_unique\n",
    "print(f'len(curr_feature_df): {len(curr_feature_df)}')\n",
    "print(f'(curr_feature_df[f\"diagnosis_class\"] != \"normal\").sum(): {(curr_feature_df[f\"diagnosis_class\"] != \"normal\").sum()}')\n",
    "\n",
    "# print(curr_ext_donor_feature_df['donor_id'].nunique())\n",
    "# print(curr_ext_donor_feature_df[curr_ext_donor_feature_df['donor_id'] == 'N198'])\n",
    "all_mean_near_neighbor_dists_without_atlas = curr_feature_df.loc[~curr_feature_df['atlas'], 'mean_near_neighbor_dist']\n",
    "normal_mean_nn_dists_without_atlas = curr_feature_df.loc[~curr_feature_df['atlas'] & (curr_feature_df['diagnosis_class'] == 'normal'), 'mean_near_neighbor_dist']\n",
    "curr_feature_df['normal'] = curr_feature_df[f'diagnosis_class'] == 'normal'\n",
    "\n",
    "normal_mean_nn_dist_without_atlas_high_quantile_val = normal_mean_nn_dists_without_atlas.quantile(nn_dist_high_quantile)\n",
    "\n",
    "if not skip_plotting:\n",
    "    plt.close('all')\n",
    "    fig, ax = plt.subplots(figsize=(3.5,3.5))\n",
    "    # sb.histplot(curr_feature_df['mean_near_neighbor_dist'], bins=30)\n",
    "    # sb.histplot(all_mean_near_neighbor_dists_without_atlas, bins=30)\n",
    "    # sb.histplot(normal_mean_nn_dists_without_atlas, bins=30)\n",
    "    sb.histplot(curr_feature_df.loc[~curr_feature_df['atlas']], x='mean_near_neighbor_dist', hue='diagnosis_class', cumulative=True, element='step', fill=False, stat='proportion', common_norm=False, bins=3000, palette=clinical_data_interface_and_utils.DIAGNOSIS_CLASS_TO_COLOR, ax=ax)\n",
    "    ax.axvline(normal_mean_nn_dist_without_atlas_high_quantile_val, color='grey', alpha=0.5)\n",
    "    ax.legend().remove()\n",
    "    supp_out_dir_path = 'temp/_fig_sup_4'\n",
    "    pathlib.Path(supp_out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "    fig.savefig(os.path.join(supp_out_dir_path, 'EDF_10A_mean_near_neighbor_dist_cum_hist.png'), dpi=300)\n",
    "    # raise\n",
    "\n",
    "feature_df_before_removing_normal = curr_feature_df.copy()\n",
    "\n",
    "if dont_show_normal:\n",
    "    curr_feature_df = curr_feature_df[curr_feature_df['diagnosis_class'] != 'normal']\n",
    "if dont_show_normal_atlas:\n",
    "    curr_feature_df = curr_feature_df[(curr_feature_df['diagnosis_class'] != 'normal') | (~curr_feature_df['atlas'])]\n",
    "\n",
    "\n",
    "assert not curr_feature_df.empty, 'filtered out all donors'\n",
    "\n",
    "observed_cell_state_names = [x for x in curr_feature_df.columns if x in all_ordered_state_names]\n",
    "missing_cell_state_names = set(observed_cell_state_names) - set(all_ordered_state_names)\n",
    "assert (not missing_cell_state_names), f'{missing_cell_state_names} are missing from all_ordered_state_names'\n",
    "\n",
    "assert set(observed_hspc_state_names) == {x for x in observed_cell_state_names if x in ordered_pb_hspc_state_names}\n",
    "df = curr_feature_df[observed_hspc_state_names + ['numbered_donor_id']].copy()\n",
    "df[log_hspc_freq_cols] = np.log2(df[observed_hspc_state_names] + epsilon_for_cell_freq)\n",
    "numbered_donor_ids_in_orig_order = df['numbered_donor_id'].to_numpy()\n",
    "numbered_donor_ids_in_orig_order_as_list = list(numbered_donor_ids_in_orig_order)\n",
    "df.set_index('numbered_donor_id', inplace=True)\n",
    "\n",
    "\n",
    "cols_for_clustering = log_hspc_freq_cols if cluster_log_freqs else observed_hspc_state_names\n",
    "df_for_plot = df[observed_hspc_state_names].copy()\n",
    "\n",
    "\n",
    "plt.close('all')\n",
    "type_to_color = mds_analysis_params.get_cell_state_to_color(add_outlier_black_color=True)\n",
    "\n",
    "figsize_x = 25\n",
    "figsize_y = 15\n",
    "\n",
    "if not skip_plotting:\n",
    "    fig = plt.figure(\n",
    "        figsize=(figsize_x, figsize_y),\n",
    "    )\n",
    "\n",
    "    num_of_rows = 8\n",
    "\n",
    "    fig_gridspec = fig.add_gridspec(\n",
    "        num_of_rows, 1, \n",
    "        height_ratios=[\n",
    "            0.2, # age, sex, diagnosis\n",
    "            \n",
    "            0.35, # space\n",
    "\n",
    "            \n",
    "            4, # HSPC composition\n",
    "            0.4 if show_donor_ids else 0.13, # space (for donor_id)\n",
    "            1, # mean_near_neighbor_dist\n",
    "            0.13, # space (for donor_id)\n",
    "            \n",
    "            0.5, # vafs, s_phase, bemp ms4a2, hla\n",
    "            \n",
    "            0.1, # space\n",
    "        ],\n",
    "    )\n",
    "\n",
    "    if 1:\n",
    "        fig.subplots_adjust(\n",
    "            top=0.97,\n",
    "            # left=0.065,\n",
    "            left=0.1,\n",
    "            # left=0.12, # fit MPN on 230820?\n",
    "            hspace=0.0000001,\n",
    "            \n",
    "            \n",
    "            bottom=0.15,\n",
    "            # wspace=0.2,\n",
    "            # right=0.78, # fit MPN on 230820?\n",
    "            right=0.88,\n",
    "        )\n",
    "    # print(fig_gridspec)\n",
    "    # print(fig_gridspec[0,0])\n",
    "    composition_ax = fig.add_subplot(fig_gridspec[2,0])\n",
    "    mean_nn_dist_ax = fig.add_subplot(fig_gridspec[4,0])\n",
    "    lower_extra_data_ax = fig.add_subplot(fig_gridspec[6,0])\n",
    "\n",
    "    # generic_utils.make_all_spines_and_x_and_y_axes_invisible(mean_nn_dist_ax)\n",
    "    # ax.get_yaxis().set_visible(True)\n",
    "    mean_nn_dist_ax.get_xaxis().set_visible(False)\n",
    "\n",
    "    lower_extra_data_ax.get_xaxis().set_visible(False)\n",
    "    lower_extra_data_ax.get_yaxis().set_visible(True)\n",
    "    lower_extra_data_ax.set_facecolor(\"pink\") # TODO: should remove?\n",
    "\n",
    "\n",
    "if order_donors_by_id:\n",
    "    assert not order_donors_by_bleeding_date_and_exp_name\n",
    "    nd_ids_sorted_by_id = sorted(numbered_donor_ids_in_orig_order)\n",
    "    assert set(numbered_donor_ids_in_orig_order) == set(nd_ids_sorted_by_id)\n",
    "    ordered_row_indices = [numbered_donor_ids_in_orig_order_as_list.index(x) for x in nd_ids_sorted_by_id]\n",
    "elif order_donors_by_bleeding_date_and_exp_name:\n",
    "    ordered_exp_names = sc_rna_seq_preprocessing_params.get_ordered_exp_names_from_df_with_bleeding_dates(curr_feature_df)\n",
    "    numbered_donor_id_to_exp_name = curr_feature_df.set_index('numbered_donor_id')['exp_name'].to_dict()\n",
    "    assert set(numbered_donor_id_to_exp_name) == set(numbered_donor_ids_in_orig_order)\n",
    "    nd_ids_sorted_by_bleeding_date_and_exp_name = sorted(\n",
    "        numbered_donor_ids_in_orig_order, key=lambda x: (ordered_exp_names.index(numbered_donor_id_to_exp_name[x]), x))\n",
    "    assert set(numbered_donor_ids_in_orig_order) == set(nd_ids_sorted_by_bleeding_date_and_exp_name)\n",
    "    ordered_row_indices = [numbered_donor_ids_in_orig_order_as_list.index(x) for x in nd_ids_sorted_by_bleeding_date_and_exp_name]\n",
    "else:\n",
    "    low_nn_dist_mask = curr_feature_df['mean_near_neighbor_dist'] <= normal_mean_nn_dist_without_atlas_high_quantile_val\n",
    "    high_nn_dist_mask = ~low_nn_dist_mask\n",
    "\n",
    "    print(f'low_nn_dist_mask.sum(): {low_nn_dist_mask.sum()}')\n",
    "    print(f'high_nn_dist_mask.sum(): {high_nn_dist_mask.sum()}')\n",
    "\n",
    "    low_nn_dist_and_any_cna_mask = curr_feature_df['any_large_dels_or_dups'] & low_nn_dist_mask\n",
    "    low_nn_dist_and_any_cna_count = low_nn_dist_and_any_cna_mask.sum()\n",
    "    low_nn_dist_and_no_cna_count = low_nn_dist_mask.sum() - low_nn_dist_and_any_cna_count\n",
    "    \n",
    "    high_nn_dist_and_any_cna_mask = curr_feature_df['any_large_dels_or_dups'] & high_nn_dist_mask\n",
    "    high_nn_dist_and_any_cna_count = high_nn_dist_and_any_cna_mask.sum()\n",
    "    high_nn_dist_and_no_cna_count = high_nn_dist_mask.sum() - high_nn_dist_and_any_cna_count\n",
    "    \n",
    "    print(f'low_nn_dist_and_any_cna_count: {low_nn_dist_and_any_cna_count}')\n",
    "    print(f'low_nn_dist_and_no_cna_count: {low_nn_dist_and_no_cna_count}')\n",
    "    print(f'high_nn_dist_and_any_cna_count: {high_nn_dist_and_any_cna_count}')\n",
    "    print(f'high_nn_dist_and_no_cna_count: {high_nn_dist_and_no_cna_count}')\n",
    "    \n",
    "    cna_fisher_result = scipy.stats.fisher_exact([\n",
    "        [low_nn_dist_and_any_cna_count, low_nn_dist_and_no_cna_count],\n",
    "        [high_nn_dist_and_any_cna_count, high_nn_dist_and_no_cna_count],\n",
    "    ], alternative='two-sided')\n",
    "    print(f'fisher result: {cna_fisher_result}')\n",
    "\n",
    "    df_for_clustering = curr_feature_df.loc[high_nn_dist_mask, cols_for_clustering]\n",
    "    df_for_clustering.index = curr_feature_df.loc[high_nn_dist_mask, 'numbered_donor_id']\n",
    "    np.random.seed(0) # added this because i seem to have got a slightly different ordering upon in two different runs, though not important. anyway, i tried 5 different seeds and didn't get different results, even though https://stackoverflow.com/questions/16016959/scipy-stats-seed/16018958#16018958 says scipy uses numpy.random to get random numbers... ah. i think i understand. when moving the code of adding CBC to feature table to after i merged the feature tables, the order of the donors changed for some reason, so now here i started from a different order, and this is why i get very slightly different clustering order...\n",
    "    linkage_mat = scipy.cluster.hierarchy.linkage(\n",
    "        df_for_clustering.to_numpy(),\n",
"        method=linkage_method, \n",
    "        metric=linkage_metric,\n",
    "        # optimal_ordering=True,\n",
    "    )\n",
    "    hclust_ordered_row_indices = scipy.cluster.hierarchy.leaves_list(linkage_mat)\n",
    "    nd_ids_sorted_by_hclust = list(df_for_clustering.iloc[hclust_ordered_row_indices].index)\n",
    "\n",
    "    df_for_clustering['hclust_clust'] = scipy.cluster.hierarchy.fcluster(\n",
    "        Z=linkage_mat,\n",
    "        t=hclust_clust_count,\n",
    "        criterion='maxclust',\n",
    "    )\n",
    "    df_for_clustering['hclust_clust'] = df_for_clustering['hclust_clust'].astype(str)\n",
    "    df_for_clustering.reset_index(inplace=True)\n",
    "    curr_feature_df = generic_utils.merge_preserving_df1_index_and_row_order(curr_feature_df, df_for_clustering[['numbered_donor_id', 'hclust_clust']], how='left')\n",
    "    curr_feature_df['hclust_clust'].fillna('-1', inplace=True)\n",
    "\n",
    "\n",
    "    curr_lymph_hspc_freq_cols = sorted(set(curr_feature_df.columns) & set([f'c_{x}' for x in mds_analysis_params.PB_LYMPHOID_HSPC_STATE_NAMES]))\n",
    "    curr_non_lymph_hspc_freq_cols = sorted(set(curr_feature_df.columns) & set([f'c_{x}' for x in mds_analysis_params.PB_NON_LYMPHOID_HSPC_STATE_NAMES]))\n",
    "    curr_feature_df['lymph_hspc_freq'] = curr_feature_df[curr_lymph_hspc_freq_cols].sum(axis=1)\n",
    "    curr_feature_df['non_lymph_hspc_freq'] = curr_feature_df[curr_non_lymph_hspc_freq_cols].sum(axis=1)\n",
    "    assert np.allclose(curr_feature_df[['lymph_hspc_freq', 'non_lymph_hspc_freq']].sum(axis=1), 1)\n",
    "    curr_feature_df['lymph_non_lymph_hspc_freq_log_ratio'] = np.log2(curr_feature_df['lymph_hspc_freq'] + epsilon_for_cell_freq) - np.log2(\n",
    "        curr_feature_df['non_lymph_hspc_freq'] + epsilon_for_cell_freq)\n",
    "    nd_ids_sorted_by_lymph_myelo_log_ratio = list(curr_feature_df.loc[low_nn_dist_mask].sort_values('lymph_non_lymph_hspc_freq_log_ratio', ascending=False)['numbered_donor_id'])\n",
    "    ordered_row_indices = [numbered_donor_ids_in_orig_order_as_list.index(x) for x in nd_ids_sorted_by_lymph_myelo_log_ratio + nd_ids_sorted_by_hclust]\n",
    "\n",
    "\n",
    "    if order_by_k_means_k is not None:\n",
    "        kmeans = sklearn.cluster.KMeans(n_clusters=order_by_k_means_k, random_state=0, init='k-means++').fit(df_for_clustering.to_numpy())\n",
    "        df_for_clustering['kmeans_cluster'] = kmeans.labels_.astype(str)\n",
    "        df_for_clustering['hclust_i'] = [nd_ids_sorted_by_hclust.index(x) for x in numbered_donor_ids_in_orig_order_as_list]\n",
    "        \n",
    "        df_for_clustering = generic_utils.merge_preserving_df1_index_and_row_order(df_for_clustering, df_for_clustering.groupby('kmeans_cluster')['c_CLP'].mean().reset_index(name='kmeans_cluster_mean_c_CLP'))\n",
    "        # assert not cluster_log_freqs\n",
    "        # df_for_clustering['CLP_and_NKTDP'] = df_for_clustering['c_CLP'] + df_for_clustering['c_NKTDP']\n",
    "        df_for_clustering = df_for_clustering.reset_index().rename(columns={'index': 'numbered_donor_id'})\n",
    "        df_for_clustering = generic_utils.merge_preserving_df1_index_and_row_order(df_for_clustering, curr_feature_df[[\n",
    "            'numbered_donor_id', \n",
    "            'DNMT3A_max_mean_VAF',\n",
    "            'TET2_max_mean_VAF',\n",
    "            'SF3B1_max_mean_VAF',\n",
    "            'max_mean_VAF_excluding_top_3',\n",
    "            'any_large_dels_or_dups',\n",
    "            'donor_age',\n",
    "        ]])\n",
    "        # nd_ids_sorted_by_kmeans = list(df_for_clustering.sort_values(['kmeans_cluster_mean_c_CLP', 'kmeans_cluster', 'CLP_and_NKTDP'])['numbered_donor_id'])\n",
    "        nd_ids_sorted_by_kmeans = list(df_for_clustering.sort_values(['kmeans_cluster_mean_c_CLP', 'kmeans_cluster', 'hclust_i',\n",
    "                                                                       'DNMT3A_max_mean_VAF',\n",
    "            'TET2_max_mean_VAF',\n",
    "            'SF3B1_max_mean_VAF',\n",
    "            'max_mean_VAF_excluding_top_3',\n",
    "            'any_large_dels_or_dups', 'donor_age'])['numbered_donor_id'])\n",
    "        \n",
    "        curr_feature_df = generic_utils.merge_preserving_df1_index_and_row_order(curr_feature_df, df_for_clustering[['numbered_donor_id', 'kmeans_cluster']], how='left')\n",
    "        assert set(numbered_donor_ids_in_orig_order) == set(nd_ids_sorted_by_kmeans)\n",
    "        ordered_row_indices = [numbered_donor_ids_in_orig_order_as_list.index(x) for x in nd_ids_sorted_by_kmeans]\n",
    "\n",
    "df_for_plot = df_for_plot.iloc[ordered_row_indices]\n",
    "\n",
    "if discard_numbered_donor_id_tech_rep_suffix:\n",
    "    df_for_plot.index = sc_rna_seq_preprocessing_params.get_numbered_donor_id_series_without_tech_rep_suffix(pd.Series(df_for_plot.index)).to_numpy() # \n",
    "\n",
    "final_donor_ids_ordered = list(numbered_donor_ids_in_orig_order[ordered_row_indices])\n",
    "\n",
    "if order_donors_by_id:\n",
    "    assert final_donor_ids_ordered == nd_ids_sorted_by_id\n",
    "elif order_donors_by_bleeding_date_and_exp_name:\n",
    "    assert final_donor_ids_ordered == nd_ids_sorted_by_bleeding_date_and_exp_name\n",
    "\n",
    "ordered_observed_hspc_states = [x for x in ordered_pb_hspc_state_names if x in df_for_plot.columns]\n",
    "\n",
    "df_for_hspc_plot = df_for_plot[ordered_observed_hspc_states]\n",
    "# assert np.isclose(df_for_hspc_plot.sum(axis=1), 1, atol=0.02).all()\n",
    "\n",
    "unexpected_sum_mask = ~np.isclose(df_for_hspc_plot.sum(axis=1), 1)\n",
    "assert not unexpected_sum_mask.any(), f'df_for_hspc_plot[unexpected_sum_mask]: {df_for_hspc_plot[unexpected_sum_mask]}'\n",
    "\n",
    "curr_feature_df_ordered = generic_utils.get_df_sorted_by_column_vals(\n",
    "    curr_feature_df, 'numbered_donor_id', final_donor_ids_ordered, left_join=True)\n",
    "# curr_ext_donor_feature_df = generic_utils.merge_preserving_df1_index_and_row_order(\n",
    "#     curr_ext_donor_feature_df, curr_ext_donor_feature_df[['numbered_donor_id', 'donor_age']].drop_duplicates(subset='numbered_donor_id'))\n",
    "\n",
    "curr_feature_df_ordered.loc[curr_feature_df_ordered['numbered_donor_id'].isin(nd_ids_sorted_by_hclust), 'composition_class'] = 'other abnorm comp'\n",
    "curr_feature_df_ordered.loc[~curr_feature_df_ordered['numbered_donor_id'].isin(nd_ids_sorted_by_hclust), 'composition_class'] = 'norm-like comp'\n",
    "\n",
    "ordered_nd_ids_without_suffix = list(curr_feature_df_ordered['numbered_donor_id_without_tech_rep_suffix'])\n",
    "for comp_class, class_nd_ids_without_suffix in comp_class_to_nd_ids_without_suffix.items():\n",
    "    indices = [ordered_nd_ids_without_suffix.index(x) for x in class_nd_ids_without_suffix]\n",
    "    min_i = min(indices)\n",
    "    max_i = max(indices)\n",
    "    assert set(indices) == set(range(min_i, max_i + 1)), f'{ordered_nd_ids_without_suffix} are not consecutive'\n",
    "\n",
    "    curr_feature_df_ordered.loc[curr_feature_df_ordered['numbered_donor_id_without_tech_rep_suffix'].isin(class_nd_ids_without_suffix), 'composition_class'] = comp_class\n",
    "\n",
    "    if not skip_plotting:\n",
    "        composition_ax.annotate('', xy=(min_i - 0.33, 1.01), xycoords=('data', 'axes fraction'), xytext=(max_i + 0.33, 1.01), \n",
    "                arrowprops=dict(arrowstyle=\"-\", linewidth=2.5, color='k'))\n",
    "        \n",
    "assert curr_feature_df_ordered['composition_class'].notna().all()\n",
    "curr_feature_df_ordered[['donor_id', 'numbered_donor_id', 'exp_name', 'composition_class']].to_csv(get_mds_params()['fig4_donor_composition_class_df_csv_file_path'], index=False)\n",
    "\n",
    "if not skip_plotting:\n",
    "\n",
    "    binary_cmap = plt.get_cmap('binary')\n",
    "    binary_cmap.set_bad(color='red', alpha=0.2)\n",
    "        \n",
    "\n",
    "    raw_extra_data_df = curr_feature_df_ordered[[\n",
    "            'DNMT3A_max_mean_VAF',\n",
    "            'TET2_max_mean_VAF',\n",
    "            'SF3B1_max_mean_VAF',\n",
    "            # 'max_mean_VAF',\n",
    "            'max_mean_VAF_excluding_top_3',\n",
    "            'any_large_dels_or_dups',\n",
    "            'diagnosis_class',\n",
    "        ]].rename(columns={\n",
    "            'log_HSPC_count': 'log(#HSPCs)', \n",
    "            'any_large_dels_or_dups': 'any_CNA',\n",
    "            'max_mean_VAF_excluding_top_3': 'other max VAF',\n",
    "            'diagnosis_class': 'diagnosis_c',\n",
    "        })\n",
    "\n",
    "    col_to_cmap_and_min_and_max = {\n",
    "        'DNMT3A_max_mean_VAF': (binary_cmap, 0, 0.5),\n",
    "        'TET2_max_mean_VAF': (binary_cmap, 0, 0.5),\n",
    "        'SF3B1_max_mean_VAF': (binary_cmap, 0, 0.5),\n",
    "        'other max VAF': (binary_cmap, 0, 0.5),\n",
    "        'any_CNA': (binary_cmap, 0, 1),\n",
    "    }\n",
    "    col_to_val_to_color = {\n",
    "        'diagnosis_c': clinical_data_interface_and_utils.DIAGNOSIS_CLASS_TO_COLOR,\n",
    "    }\n",
    "\n",
    "    def plot_col_color(ax, cols):\n",
    "        curr_raw_extra_data_df = raw_extra_data_df[cols].copy()\n",
    "        color_arr = generic_utils.get_color_arr(\n",
    "            curr_raw_extra_data_df, allow_nans=True, \n",
    "            col_to_cmap_and_min_and_max={k: v for k, v in col_to_cmap_and_min_and_max.items() if k in cols},\n",
    "            col_to_val_to_color={k: v for k, v in col_to_val_to_color.items() if k in cols},\n",
    "        ).transpose((1, 0, 2))\n",
    "        ax.imshow(color_arr, aspect='auto')\n",
    "        ax.set_yticks(range(len(curr_raw_extra_data_df.columns)))\n",
    "        ax.set_yticklabels(curr_raw_extra_data_df.columns)\n",
    "\n",
    "    assert set(lower_ax_cols) == set(raw_extra_data_df.columns)\n",
    "    plot_col_color(lower_extra_data_ax, lower_ax_cols)\n",
    "\n",
    "    # df_for_plot.index = [(f'norm_{x}' if x in healthy_donor_ids else x) for x in df_for_plot.index]\n",
    "\n",
    "    assert all(x.startswith('c_') for x in ordered_pb_hspc_state_names)\n",
    "    df_for_hspc_plot = df_for_hspc_plot.copy()\n",
    "    df_for_hspc_plot.rename(columns={x: x[2:] for x in ordered_pb_hspc_state_names}, inplace=True)\n",
    "\n",
    "    # make sure the stacked bar composition shows at most a single sample per donor\n",
    "    df_for_sanity_check = generic_utils.merge_preserving_df1_index_and_row_order(\n",
    "        pd.DataFrame({'numbered_donor_id_without_tech_rep_suffix': df_for_hspc_plot.index}), \n",
    "        all_ext_feature_df[['numbered_donor_id_without_tech_rep_suffix', 'donor_id']].drop_duplicates(),\n",
    "    )\n",
    "    assert df_for_sanity_check['donor_id'].is_unique\n",
    "\n",
    "    print(f'len(df_for_hspc_plot): {len(df_for_hspc_plot)}')\n",
    "\n",
    "    df_for_hspc_plot.plot(\n",
    "        kind=\"bar\",\n",
    "        ax=composition_ax,\n",
    "        stacked=True,\n",
    "        color=[\n",
    "            type_to_color[x]\n",
    "            for x in list(df_for_hspc_plot)\n",
    "        ],\n",
    "        # legend=False,\n",
    "    )\n",
    "    composition_ax.set_xlabel(None)\n",
    "    if not show_donor_ids:\n",
    "        composition_ax.get_xaxis().set_visible(False)\n",
    "    composition_ax.set_ylim(0,1.001)\n",
    "\n",
    "    handles, labels = composition_ax.get_legend_handles_labels()\n",
    "    legend_handles = handles\n",
    "    legend_labels = labels\n",
    "\n",
    "    legend_ax = composition_ax\n",
    "    legend_ax.legend(legend_handles[::-1], legend_labels[::-1], fontsize='x-large')\n",
    "    sb.move_legend(legend_ax, \"upper left\", bbox_to_anchor=(1, 1), fontsize='x-large')\n",
    "\n",
    "\n",
    "    min_nn_dist = curr_feature_df_ordered['mean_near_neighbor_dist'].min()\n",
    "    max_nn_dist = curr_feature_df_ordered['mean_near_neighbor_dist'].max()\n",
    "    norm_normal_mean_nn_dist_without_atlas_high_quantile_val = (normal_mean_nn_dist_without_atlas_high_quantile_val - min_nn_dist) / (max_nn_dist - min_nn_dist)\n",
    "    print(f'norm_normal_mean_nn_dist_without_atlas_high_quantile_val: {norm_normal_mean_nn_dist_without_atlas_high_quantile_val}')\n",
    "\n",
    "    mean_nn_dist_cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\n",
    "    '', [(0, 'grey'), (norm_normal_mean_nn_dist_without_atlas_high_quantile_val, 'grey'), (1, 'red')])\n",
    "    # curr_feature_df_ordered['norm_mean_nn_dist'] = generic_utils.normalize_df_columns(\n",
    "    #     curr_feature_df_ordered, column_names=['mean_near_neighbor_dist'], \n",
    "    #     col_to_min_and_max={'mean_near_neighbor_dist': (min_nn_dist, max_nn_dist)},\n",
    "    # )['mean_near_neighbor_dist']\n",
    "\n",
    "    curr_feature_df_ordered['mean_nn_dist_color'] = generic_utils.get_color_df(\n",
    "        curr_feature_df_ordered[['mean_near_neighbor_dist']], \n",
    "        col_to_cmap_and_min_and_max={'mean_near_neighbor_dist': (mean_nn_dist_cmap, min_nn_dist, max_nn_dist)})['mean_near_neighbor_dist']\n",
    "    mean_nn_dist_palette = generic_utils.get_dict_mapping_one_df_column_to_other(curr_feature_df_ordered, 'numbered_donor_id', 'mean_nn_dist_color')\n",
    "    # raise\n",
    "    sb.barplot(data=curr_feature_df_ordered, x='numbered_donor_id', y='mean_near_neighbor_dist', hue='numbered_donor_id', palette=mean_nn_dist_palette, ax=mean_nn_dist_ax, dodge=False)\n",
    "    mean_nn_dist_ax.get_legend().remove()\n",
    "\n",
    "    ordered_by_k_means_repr = '' if (order_by_k_means_k is None) else f'_kmeans{order_by_k_means_k}'\n",
    "    dont_show_normal_atlas_repr = f'_no_normal_atlas' if dont_show_normal_atlas else ''        \n",
    "    dont_show_normal_repr = f'_no_normal' if dont_show_normal else ''        \n",
    "    nn_dist_high_quantile_repr = f'_{nn_dist_high_quantile}'\n",
    "    fig_file_name = f'fig_4BCD_{ordered_by_k_means_repr}{dont_show_normal_atlas_repr}{dont_show_normal_repr}{nn_dist_high_quantile_repr}.png'\n",
    "    fig.savefig(\n",
    "        os.path.join(out_dir_path, fig_file_name), \n",
    "        dpi=300,\n",
    "    )\n",
    "    plt.close('all')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "horizontal = False\n",
    "horizontal = True\n",
    "orientation = 'horizontal' if horizontal else 'vertical'\n",
    "\n",
    "mean_nn_dist_cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\n",
    "    '', [(0, 'grey'), (norm_normal_mean_nn_dist_without_atlas_high_quantile_val, 'grey'), (1, 'red')])\n",
    "\n",
    "print(normal_mean_nn_dist_without_atlas_high_quantile_val)\n",
    "plt.close('all')\n",
    "fig, ax = plt.subplots(figsize=(3, 1) if horizontal else (1, 3))\n",
    "if horizontal:\n",
    "    fig.subplots_adjust(bottom=0.5)\n",
    "else:\n",
    "    fig.subplots_adjust(right=0.5)\n",
    "\n",
    "# cmap = matplotlib.cm.cool\n",
    "norm = matplotlib.colors.Normalize(vmin=0, vmax=max_nn_dist)\n",
    "\n",
    "cb1 = matplotlib.colorbar.ColorbarBase(ax, cmap=mean_nn_dist_cmap,\n",
    "                                norm=norm,\n",
    "                                orientation=orientation)\n",
    "cb1.set_label('composition abnormality score')\n",
    "\n",
    "fig.savefig(\n",
    "    os.path.join(out_dir_path, f'FIG_4C_comp_abnorm_score_colorbar_{orientation}.png'), \n",
    "    dpi=300,\n",
    ")\n",
    "\n",
    "plt.close('all')\n",
    "fig, ax = plt.subplots(figsize=(3, 1) if horizontal else (1, 3))\n",
    "if horizontal:\n",
    "    fig.subplots_adjust(bottom=0.5)\n",
    "else:\n",
    "    fig.subplots_adjust(right=0.5)\n",
    "\n",
    "# cmap = matplotlib.cm.cool\n",
    "norm = matplotlib.colors.Normalize(vmin=0, vmax=50)\n",
    "\n",
    "cb1 = matplotlib.colorbar.ColorbarBase(ax, cmap=binary_cmap,\n",
    "                                norm=norm,\n",
    "                                orientation=orientation)\n",
    "cb1.set_label('VAF')\n",
    "\n",
    "fig.savefig(\n",
    "    os.path.join(out_dir_path, f'FIG_4D_vaf_colorbar_{orientation}.png'), \n",
    "    dpi=300,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig_sup_4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "for ordered_numbered_donor_ids_without_suffix in [\n",
    "    [\n",
    "        'N192_1',\n",
    "        'N192_2',\n",
    "    ],\n",
    "    [\n",
    "        'N235_1',\n",
    "        'N235_2',\n",
    "    ],\n",
    "    [\n",
    "        'N281_1',\n",
    "        'N281_2',\n",
    "    ],\n",
    "    [\n",
    "        'N204_1',\n",
    "        'N204_2',\n",
    "    ],\n",
    "    [\n",
    "        'N226_1',\n",
    "        'N226_2',\n",
    "    ],\n",
    "    [\n",
    "        'N224_1',\n",
    "        'N224_2',\n",
    "    ],\n",
    "    # 'N74',\n",
    "]:\n",
    "    epsilon_for_cell_freq = get_mds_params()['epsilon_for_donor_log_c_state_freq']\n",
    "\n",
    "    all_ordered_state_names = [f'c_{x}' for x in mds_analysis_params.ORDERED_CELL_STATE_NAMES]\n",
    "    ordered_pb_hspc_state_names = [f'c_{x}' for x in mds_analysis_params.ORDERED_PB_HSPC_STATE_NAMES]\n",
    "\n",
    "    all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "    curr_feature_df = all_ext_feature_df.copy()\n",
    "    curr_feature_df = curr_feature_df[\n",
    "        (curr_feature_df['final_c_hspc_count'] >= 450) &\n",
    "        ~curr_feature_df['biased_composition_due_to_discarded_low_umi_count_barcodes']\n",
    "    ].copy()\n",
    "    \n",
    "    curr_feature_df['diagnosis_class'] = curr_feature_df['diagnosis_class'].astype(str)\n",
    "\n",
    "    observed_hspc_state_names = sorted(set(curr_feature_df.columns) & set(ordered_pb_hspc_state_names))\n",
    "    log_hspc_freq_cols = [f'log_{x}' for x in observed_hspc_state_names]\n",
    "    curr_feature_df[log_hspc_freq_cols] = np.log2(curr_feature_df[observed_hspc_state_names] + epsilon_for_cell_freq)\n",
    "\n",
    "    unexpected_diagnosis_mask = ~curr_feature_df['diagnosis_class'].isin(['MDS', 'MDS/MPN', 'cytopenia', 'normal'])\n",
    "    assert not unexpected_diagnosis_mask.any()\n",
    "\n",
    "    # curr_feature_df = curr_feature_df[curr_feature_df['in_mds_cyto_groups_analysis']]\n",
    "    curr_feature_df = curr_feature_df[curr_feature_df['numbered_donor_id'].isin(\n",
    "        sc_rna_seq_preprocessing_params.get_best_or_random_numbered_donor_id_per_donor_or_donor_sample(curr_feature_df))]\n",
    "\n",
    "    # curr_feature_df = curr_feature_df[curr_feature_df['diagnosis_class'] == 'normal']\n",
    "\n",
    "    observed_cell_state_names = [x for x in curr_feature_df.columns if x in all_ordered_state_names]\n",
    "    missing_cell_state_names = set(observed_cell_state_names) - set(all_ordered_state_names)\n",
    "    assert (not missing_cell_state_names), f'{missing_cell_state_names} are missing from all_ordered_state_names'\n",
    "\n",
    "    assert set(observed_hspc_state_names) == {x for x in observed_cell_state_names if x in ordered_pb_hspc_state_names}\n",
    "    df = curr_feature_df[observed_hspc_state_names + ['numbered_donor_id']].copy()\n",
    "    df[log_hspc_freq_cols] = np.log2(df[observed_hspc_state_names] + epsilon_for_cell_freq)\n",
    "    df.set_index('numbered_donor_id', inplace=True)\n",
    "    df_for_plot = df[observed_hspc_state_names].copy()\n",
    "\n",
    "\n",
    "    plt.close('all')\n",
    "    type_to_color = mds_analysis_params.get_cell_state_to_color(add_outlier_black_color=True)\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(2, 4))\n",
    "\n",
    "    df_for_plot.index = sc_rna_seq_preprocessing_params.get_numbered_donor_id_series_without_tech_rep_suffix(pd.Series(df_for_plot.index)).to_numpy() # \n",
    "\n",
    "    ordered_observed_hspc_states = [x for x in ordered_pb_hspc_state_names if x in df_for_plot.columns]\n",
    "    df_for_hspc_plot = df_for_plot[ordered_observed_hspc_states]\n",
    "    print(df_for_hspc_plot.index)\n",
    "    df_for_hspc_plot = df_for_hspc_plot.loc[ordered_numbered_donor_ids_without_suffix]\n",
    "    # df_for_hspc_plot = df_for_hspc_plot.loc[sorted(df_for_hspc_plot.index)]\n",
    "\n",
    "    unexpected_sum_mask = ~np.isclose(df_for_hspc_plot.sum(axis=1), 1)\n",
    "    assert not unexpected_sum_mask.any(), f'df_for_hspc_plot[unexpected_sum_mask]: {df_for_hspc_plot[unexpected_sum_mask]}'\n",
    "\n",
    "    df_for_hspc_plot = df_for_hspc_plot.copy()\n",
    "    df_for_hspc_plot.rename(columns={x: x[2:] for x in ordered_pb_hspc_state_names}, inplace=True)\n",
    "\n",
    "    df_for_hspc_plot.plot(\n",
    "        kind=\"bar\",\n",
    "        ax=ax,\n",
    "        stacked=True,\n",
    "        color=[\n",
    "            type_to_color[x]\n",
    "            for x in list(df_for_hspc_plot)\n",
    "        ],\n",
    "        legend=False,\n",
    "    )\n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_ylim(0,1.001)\n",
    "    generic_utils.make_all_spines_and_x_and_y_axes_invisible(ax)\n",
    "    ax.get_xaxis().set_visible(True)\n",
    "    fig.tight_layout()\n",
    "\n",
    "    fig.savefig(\n",
    "        os.path.join(out_dir_path, f'EDF_12C_bio_rep_compos_stacked_{ordered_numbered_donor_ids_without_suffix[0]}.png'), \n",
    "        dpi=300,\n",
    "    )\n",
    "    plt.close('all')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "sorted_diagnosis_and_color = [(x, clinical_data_interface_and_utils.DIAGNOSIS_CLASS_TO_COLOR[x]) for x in ['normal', 'cytopenia', 'MDS', 'MDS/MPN']]\n",
    "\n",
    "plt.close('all')\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.legend(handles=[matplotlib.patches.Patch(color=x[1], label=x[0]) for x in sorted_diagnosis_and_color], fontsize='x-large')\n",
    "\n",
    "fig.savefig(\n",
    "    os.path.join(out_dir_path, f'patch_diagnosis_legend.png'), \n",
    "    dpi=300,\n",
    ")\n",
    "\n",
    "plt.close('all')\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "for x in sorted_diagnosis_and_color:\n",
    "    ax.scatter([0.5], [0.5], s=60, color=x[1], label=x[0])\n",
    "\n",
    "ax.legend(fontsize='large')\n",
    "sb.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1), fontsize='large')\n",
    "fig.tight_layout()\n",
    "fig.savefig(\n",
    "    os.path.join(out_dir_path, f'circle_diagnosis_legend.png'), \n",
    "    dpi=300,\n",
    ")\n",
    "\n",
    "plt.close('all')\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "for x in sorted_diagnosis_and_color:\n",
    "    ax.plot([0.5], [0.5], color=x[1], label=x[0])\n",
    "\n",
    "ax.legend(fontsize='large')\n",
    "sb.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1), fontsize='large')\n",
    "fig.tight_layout()\n",
    "fig.savefig(\n",
    "    os.path.join(out_dir_path, f'line_diagnosis_legend.png'), \n",
    "    dpi=300,\n",
    ")\n",
    "\n",
    "plt.close('all')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10,10))\n",
    "\n",
    "if 'n2_c_ad' not in locals():\n",
    "    raise RuntimeError('run code above to get n2_c_ad with updated c_states (e.g., without CFD_tryptase_GMP-L) - need to call mc_utils.add_c_state_and_mc_c_state_stats()')\n",
    "\n",
    "\n",
    "assert set(n2_c_ad.obs['c_state'].astype(str)) <= set(mds_analysis_params.ORDERED_CELL_STATE_NAMES)\n",
    "\n",
    "ordered_c_states = [x for x in mds_analysis_params.ORDERED_PB_HSPC_STATE_NAMES if x in n2_c_ad.obs['c_state'].unique()]\n",
    "ordered_c_states = [x for x in mds_analysis_params.ORDERED_CELL_STATE_NAMES if x in n2_c_ad.obs['c_state'].unique()]\n",
    "state_color_df = pd.DataFrame({'state': ordered_c_states})\n",
    "state_color_df = generic_utils.merge_preserving_df1_index_and_row_order(state_color_df, pd.read_csv(get_mds_params()['cell_type_colors_csv_file_path']).rename(columns={'cell_type': 'state'}))\n",
    "for x in state_color_df.to_records(index=False).tolist()[::-1]:\n",
    "    ax.scatter([0.5], [0.5], s=40, color=x[1], label=x[0])\n",
    "\n",
    "ax.legend(fontsize='large')\n",
    "sb.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1), fontsize='large')\n",
    "fig.tight_layout()\n",
    "fig.savefig(\n",
    "    os.path.join(out_dir_path, f'cell_state_circle_legend.png'), \n",
    "    dpi=300,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "krusk_with_normal = True\n",
    "krusk_with_normal = False\n",
    "\n",
    "final_feature_names_and_types_df = pd.read_csv(get_mds_params()['final_feature_names_and_types_df_csv_file_path'])\n",
    "\n",
    "list_of_max_pval_and_star_count = [\n",
    "    (0.001, 3),\n",
    "    (0.01, 2),\n",
    "    (0.05, 1),\n",
    "]\n",
    "\n",
    "list_of_attrs_to_plot = [\n",
    "    [\n",
    "        'mebemp_l_hla_sig_MEBEMP-L_0.5',\n",
    "        # 'prss2_sig_CLP_0.5', # nah, though maybe maybe interesting? anyway very nice in bio reps.\n",
    "        'bemp_cd74_sig_BEMP_0.5',\n",
    "        's_phase_sig_MEBEMP-L_0.5',\n",
    "        # 'gmp_l_elane_sig_GMP-L_0.5', # nah, though maybe maybe interesting? anyway looks ok in bio reps if considering only normal. N280 is an outlier there (and she was treated was EPO)\n",
    "    ],\n",
    "    [\n",
    "        'donor_age',\n",
    "        'Hemoglobin (g/dl)',\n",
    "        'RBC (10^6/microliter)',\n",
    "    ],\n",
    "    [\n",
    "        'RDW (%)',\n",
    "        'MCV (fL)',\n",
    "        'Hematocrit (%)',\n",
    "    ],\n",
    "    [\n",
    "        'Neutro#',\n",
    "        'Mono%',\n",
    "        'WBC (10^3/microliter)',\n",
    "        # 'mean_near_neighbor_dist', # TODO: maybe not show this? it is biased for atlas, because one neighbor is at distance=0... so if showing this, should not show atlas donors\n",
    "    ],\n",
    "    [\n",
    "        ###'Lympho%',\n",
    "        'Lympho#', # looks better than Lympho% here.\n",
    "        'Platelets (10^3/microliter)',\n",
    "        'log_c_clp_e',\n",
    "    ],\n",
    "]\n",
    "all_attrs_to_plot = list(itertools.chain(*list_of_attrs_to_plot))\n",
    "\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "curr_feature_df = all_ext_feature_df[all_ext_feature_df['in_mds_cyto_groups_analysis']]\n",
    "\n",
    "curr_feature_df = generic_utils.merge_preserving_df1_index_and_row_order(curr_feature_df, pd.read_csv(get_mds_params()['fig4_donor_composition_class_df_csv_file_path'])[['numbered_donor_id', 'composition_class']], how='left')\n",
    "assert (curr_feature_df.loc[curr_feature_df['composition_class'].isna(), 'diagnosis_class'] == 'normal').all()\n",
    "curr_feature_df['composition_class'].fillna('normal diagnosis', inplace=True)\n",
    "\n",
    "\n",
    "curr_feature_df['rough_composition_class'] = curr_feature_df['composition_class'].replace({\n",
    "    'high CLP low MEBEMP': 'high CLP',\n",
    "    'high BEMP': 'other abnorm comp',\n",
    "    'only HSC/MPP': 'other abnorm comp',\n",
    "})\n",
    "curr_feature_df['rough_class_and_sex'] = curr_feature_df['rough_composition_class'].astype(str) + ' ' + curr_feature_df['donor_sex']\n",
    "ordered_classes = [\n",
    "    'normal diagnosis',\n",
    "    'norm-like comp',\n",
    "    'high CLP',\n",
    "    # 'high CLP low MEBEMP',\n",
    "    # 'high BEMP',\n",
    "    # 'only HSC/MPP',\n",
    "    'high GMP',\n",
    "    'other abnorm comp',\n",
    "]\n",
    "ref_class = 'normal diagnosis'\n",
    "class_and_sex_to_ref_class_and_sex = {}\n",
    "for clas in ordered_classes:\n",
    "    for sex in ('male', 'female'):\n",
    "        clas_and_sex = f'{clas} {sex}'\n",
    "        if clas != ref_class:\n",
    "            class_and_sex_to_ref_class_and_sex[clas_and_sex] = f'{ref_class} {sex}'\n",
    "\n",
    "ordered_class_and_sex = [\n",
    "    'normal diagnosis male',\n",
    "    'normal diagnosis female',\n",
    "    'norm-like comp male',\n",
    "    'norm-like comp female',\n",
    "    'high CLP male',\n",
    "    'high CLP female',\n",
    "    'high GMP male',\n",
    "    'high GMP female',\n",
    "    'other abnorm comp male',\n",
    "    'other abnorm comp female',\n",
    "]\n",
    "class_and_sex_to_pos_in_stripplot = {\n",
    "    'normal diagnosis male': 0,\n",
    "    'normal diagnosis female': 1,\n",
    "    'norm-like comp male': 2.5,\n",
    "    'norm-like comp female': 3.5,\n",
    "    'high CLP male': 5,\n",
    "    'high CLP female': 6,\n",
    "    'high GMP male': 7.5,\n",
    "    'high GMP female': 8.5,\n",
    "    'other abnorm comp male': 10,\n",
    "    'other abnorm comp female': 11,\n",
    "}\n",
    "male_female_vline_xs = [0.5, 3, 5.5, 8, 10.5]\n",
    "class_vline_xs = [1.75, 4.25, 6.75, 9.25]\n",
    "assert pd.Series([class_and_sex_to_pos_in_stripplot[x] for x in ordered_class_and_sex]).is_monotonic_increasing\n",
    "\n",
    "\n",
    "\n",
    "final_feature_name_to_type = generic_utils.get_dict_mapping_one_df_column_to_other(final_feature_names_and_types_df, 'name', 'type')\n",
    "\n",
    "flat_dicts = []\n",
    "for class_col in ['composition_class', 'rough_composition_class']:\n",
    "    print(class_col)\n",
    "    for feature in sorted(set(final_feature_names_and_types_df['name']) - {'donor_sex'}):\n",
    "        # print(feature)\n",
    "        curr_df = curr_feature_df[curr_feature_df[feature].notna()]\n",
    "        krusk_df = curr_df\n",
    "        if not krusk_with_normal:\n",
    "            krusk_df = krusk_df[krusk_df['diagnosis_class'] != 'normal']\n",
    "        krusk_stat, krusk_pval = scipy.stats.kruskal(*krusk_df.groupby(class_col)[feature].apply(list))\n",
    "        flat_dicts.append({'class_col': class_col, 'feature': feature, 'feature_type': final_feature_name_to_type[feature], 'test_type': 'krusk', 'krusk_stat': krusk_stat, 'pval': krusk_pval})\n",
    "        if class_col == 'rough_composition_class':\n",
    "            for clas_and_sex, ref_clas_and_sex in class_and_sex_to_ref_class_and_sex.items():\n",
    "                clas_vals = curr_df.loc[curr_df['rough_class_and_sex'] == clas_and_sex, feature]\n",
    "                ref_vals = curr_df.loc[curr_df['rough_class_and_sex'] == ref_clas_and_sex, feature]\n",
    "                if (len(clas_vals) >= 1) and (len(ref_vals) >= 1):\n",
    "                    mw_test_res = generic_utils.perform_mw_test(\n",
    "                        clas_vals,\n",
    "                        ref_vals,\n",
    "                        alternative='two-sided',\n",
    "                    )\n",
    "                    flat_dicts.append({'class_col': class_col, 'clas_and_sex': clas_and_sex, 'feature': feature, 'feature_type': final_feature_name_to_type[feature], 'test_type': 'mw', 'pval': mw_test_res['pvalue']})\n",
    "\n",
    "diff_feature_across_our_groups_df = pd.DataFrame(flat_dicts)\n",
    "diff_feature_across_our_groups_df.sort_values('pval', inplace=True)\n",
    "mw_pval_mask = diff_feature_across_our_groups_df['test_type'] == 'mw'\n",
    "diff_feature_across_our_groups_df.loc[mw_pval_mask, 'mw_bh_adjusted_pval'] = scipy.stats.false_discovery_control(\n",
    "    diff_feature_across_our_groups_df.loc[mw_pval_mask, 'pval'])\n",
    "\n",
    "\n",
    "gene_prog_kruskal_rough_comp_df = diff_feature_across_our_groups_df[\n",
    "    diff_feature_across_our_groups_df['feature_type'].isin(['gene_prog_excluding_mhc_ii', 'mhc_ii_gene_prog']) & \n",
    "    diff_feature_across_our_groups_df['class_col'].isin(['rough_composition_class']) & \n",
    "    diff_feature_across_our_groups_df['test_type'].isin(['krusk'])\n",
    "]\n",
    "\n",
    "within_state_sig_df = mds_analysis_params.get_within_state_sig_df()\n",
    "minimal_gene_prog_kruskal_rough_comp_df = generic_utils.merge_preserving_df1_index_and_row_order(gene_prog_kruskal_rough_comp_df.drop(columns=['mw_bh_adjusted_pval', 'clas_and_sex', 'krusk_stat', 'feature_type', 'test_type', 'class_col']), within_state_sig_df)\n",
    "minimal_gene_prog_kruskal_rough_comp_df['signature name'] = minimal_gene_prog_kruskal_rough_comp_df['sig_name'].replace(\n",
    "    get_mds_params()['sig_name_to_name_in_paper'])\n",
    "minimal_gene_prog_kruskal_rough_comp_df.rename(columns={'state': 'state within which signature was evaluated', 'pval': 'kruskal pvalue'}, inplace=True)\n",
    "\n",
    "\n",
    "supp_table_out_dir_path = 'temp/__rejected_sup_tables'\n",
    "pathlib.Path(supp_table_out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "minimal_gene_prog_kruskal_rough_comp_df[['signature name', 'state within which signature was evaluated', 'kruskal pvalue']].to_csv(os.path.join(supp_table_out_dir_path, 'within_state_gene_signature_kruskal_across_fig4_groups.csv'), index=False)\n",
    "\n",
    "if 0:\n",
    "    diff_feature_across_our_groups_df[~diff_feature_across_our_groups_df['plotted_in_fig4_or_sup'] & ~diff_feature_across_our_groups_df['feature_type'].isin(['log_c_state_freq'])].head(40)\n",
    "    diff_feature_across_our_groups_df[diff_feature_across_our_groups_df['plotted_in_fig4_or_sup']].head(40)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "curr_feature_df['rough_class_and_sex_pos_in_stripplot'] = curr_feature_df['rough_class_and_sex'].map(class_and_sex_to_pos_in_stripplot)\n",
    "np.random.seed(0)\n",
    "curr_feature_df['jittered_rough_class_and_sex_pos_in_stripplot'] = curr_feature_df['rough_class_and_sex_pos_in_stripplot'] + np.random.uniform(-0.3, 0.3, len(curr_feature_df))\n",
    "\n",
    "\n",
    "print(f'len(curr_feature_df): {len(curr_feature_df)}')\n",
    "for fig_i, attrs_to_plot in enumerate(list_of_attrs_to_plot):\n",
    "\n",
    "    plt.close('all')\n",
    "    fig, axes = plt.subplots(ncols=len(attrs_to_plot), figsize=(13, 2.2))\n",
    "    # fig_gridspec = fig.add_gridspec(\n",
    "    #     num_of_rows, 1, \n",
    "    #     height_ratios=[\n",
    "    #         0.2, # age, sex, diagnosis\n",
    "            \n",
    "    #         0.35, # space\n",
    "\n",
    "            \n",
    "    #         4, # HSPC composition\n",
    "    #         0.4 if show_donor_ids else 0.13, # space (for donor_id)\n",
    "    #         1, # mean_near_neighbor_dist\n",
    "    #         0.13, # space (for donor_id)\n",
    "            \n",
    "    #         0.5, # vafs, s_phase, bemp ms4a2, hla\n",
    "            \n",
    "    #         0.1, # space\n",
    "    #     ],\n",
    "    # )\n",
    "\n",
    "    if 1:\n",
    "        fig.subplots_adjust(\n",
    "            top=0.9,\n",
    "            # left=0.065,\n",
    "            left=0.05,\n",
    "            # left=0.12, # fit MPN on 230820?\n",
    "            # hspace=0.0000001,\n",
    "            \n",
    "            \n",
    "            bottom=0.05,\n",
    "            # wspace=0.2,\n",
    "            # right=0.78, # fit MPN on 230820?\n",
    "            right=0.99,\n",
    "        )\n",
    "\n",
    "    for i, (ax, attr) in enumerate(zip(axes, attrs_to_plot)):\n",
    "        curr_curr_feature_df = curr_feature_df[curr_feature_df[attr].notna()]\n",
    "        curr_observed_class_and_sex = set(curr_curr_feature_df['rough_class_and_sex'].tolist())\n",
    "        curr_ordered_class_and_sex = [x for x in ordered_class_and_sex if x in curr_observed_class_and_sex]\n",
    "        assert curr_ordered_class_and_sex == ordered_class_and_sex\n",
    "                \n",
    "                # raise\n",
    "        assert curr_curr_feature_df['donor_id'].is_unique\n",
    "        \n",
    "        if 0:\n",
    "            sb.stripplot(\n",
    "                data=curr_curr_feature_df, \n",
    "                # x='rough_class_and_sex', \n",
    "                x='rough_class_and_sex_pos_in_stripplot', \n",
    "                dodge=False,\n",
    "                jitter=0.3,\n",
    "                linewidth=0.5,\n",
    "                # order=curr_ordered_class_and_sex,\n",
    "                y=attr, \n",
    "                ax=ax,\n",
    "                hue='diagnosis_class', \n",
    "                palette=clinical_data_interface_and_utils.DIAGNOSIS_CLASS_TO_COLOR,\n",
    "                s=4,\n",
    "            )\n",
    "        else:\n",
    "            \n",
    "            sb.scatterplot(\n",
    "                data=curr_curr_feature_df, \n",
    "                x='jittered_rough_class_and_sex_pos_in_stripplot', \n",
    "                y=attr, \n",
    "                ax=ax,\n",
    "                hue='diagnosis_class', \n",
    "                palette=clinical_data_interface_and_utils.DIAGNOSIS_CLASS_TO_COLOR,\n",
    "                edgecolor='k',\n",
    "                s=20,\n",
    "            )\n",
    "            ax.get_xaxis().set_visible(False)\n",
    "\n",
    "        orig_y_lim = ax.get_ylim()\n",
    "        y_lim_range = orig_y_lim[1] - orig_y_lim[0]\n",
    "        ax.set_ylim(orig_y_lim[0], orig_y_lim[1] + 0.07 * y_lim_range)\n",
    "        \n",
    "        for clas_and_sex in curr_ordered_class_and_sex:\n",
    "            if clas_and_sex in class_and_sex_to_ref_class_and_sex:\n",
    "                mw_test_res = generic_utils.perform_mw_test(\n",
    "                    curr_curr_feature_df.loc[curr_curr_feature_df['rough_class_and_sex'] == clas_and_sex, attr],\n",
    "                    curr_curr_feature_df.loc[curr_curr_feature_df['rough_class_and_sex'] == class_and_sex_to_ref_class_and_sex[clas_and_sex], attr],\n",
    "                    alternative='two-sided',\n",
    "                )\n",
    "                curr_pval = mw_test_res['pvalue']\n",
    "\n",
    "                row_in_diff_feature_df = diff_feature_across_our_groups_df[\n",
    "                    (diff_feature_across_our_groups_df['test_type'] == 'mw') &\n",
    "                    (diff_feature_across_our_groups_df['feature'] == attr) & \n",
    "                    (diff_feature_across_our_groups_df['clas_and_sex'] == clas_and_sex) \n",
    "                ]\n",
    "                assert len(row_in_diff_feature_df) == 1\n",
    "                row_in_diff_feature_df = row_in_diff_feature_df.iloc[0]\n",
    "                assert np.isclose(row_in_diff_feature_df['pval'], curr_pval)\n",
    "                assert np.isclose(row_in_diff_feature_df['pval'], curr_pval)\n",
    "                bh_adjusted_pval = row_in_diff_feature_df['mw_bh_adjusted_pval']\n",
    "\n",
    "                # raise\n",
    "                # assert  == final_feature_name_to_type[attr]\n",
    "\n",
    "                for max_pval, star_count in list_of_max_pval_and_star_count:\n",
    "                    if bh_adjusted_pval < max_pval:\n",
    "                        x_pos = class_and_sex_to_pos_in_stripplot[clas_and_sex]\n",
    "                        x_pos = x_pos + 0.125 * (1 if clas_and_sex.endswith('female') else -1) \n",
    "                        ax.annotate(\n",
    "                            # '$' + '*' * star_count + '$',\n",
    "                            # '$*$' * star_count,\n",
    "                            '*' * star_count,\n",
    "                            # '$' + '\\\\star' * star_count + '$', # looks bad for some reason\n",
    "                            xy=(x_pos, 0.944), xycoords=('data', 'axes fraction'), color='k', ha='center', va='center', fontsize='large',\n",
    "                        )\n",
    "                        \n",
    "                        break\n",
    "\n",
    "        \n",
    "        \n",
    "        ax.set_ylabel(None)\n",
    "        ax.set_xlabel(None)\n",
    "        ax.set_title(attr)\n",
    "        if 0:\n",
    "            ax.set_xticklabels(\n",
    "                ax.get_xticklabels(),\n",
    "                rotation=90, ha='center', \n",
    "                # rotation_mode='anchor', \n",
    "                # fontsize='small',\n",
    "            )\n",
    "        for x in male_female_vline_xs:\n",
    "            ax.axvline(x, color='grey', alpha=0.5, linestyle='--')\n",
    "        for x in class_vline_xs:\n",
    "            ax.axvline(x, color='black', alpha=0.9, linestyle='-')\n",
    "\n",
    "        # if plot_legend:\n",
    "        #     legend = ax.legend(fontsize='x-large')\n",
    "        #     legend.get_frame().set_alpha(None)\n",
    "        ax.get_legend().remove()\n",
    "    # fig.tight_layout()\n",
    "    fig.savefig(\n",
    "        os.path.join(out_dir_path, f'fig_4EG_EDF_10C_sig_stripplots{fig_i}.png'), \n",
    "        dpi=300,\n",
    "    )\n",
    "    # break\n",
    "\n",
    "print('top mw gene sigs:')\n",
    "print(diff_feature_across_our_groups_df[mw_pval_mask & diff_feature_across_our_groups_df['feature_type'].isin(['gene_prog_excluding_mhc_ii', 'mhc_ii_gene_prog'])].head(15))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# total donor counts\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "\n",
    "latest_diagnosis_df = all_ext_feature_df[['donor_id', 'latest_diagnosis_class']].drop_duplicates()\n",
    "print('all latest diagnosis counts')\n",
    "print(latest_diagnosis_df['latest_diagnosis_class'].value_counts())\n",
    "# latest_diagnosis_df.to_csv('donor_id_latest_diagnosis_class.csv', index=False)\n",
    "\n",
    "print('fig2b latest diagnosis counts')\n",
    "print(all_ext_feature_df[all_ext_feature_df['in_mds_cyto_groups_analysis']]['latest_diagnosis_class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bio reps\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "curr_feature_df = all_ext_feature_df[all_ext_feature_df['in_mds_cyto_groups_analysis']]\n",
    "print(curr_feature_df['diagnosis_class'].value_counts())\n",
    "\n",
    "all_ext_feature_df['bleeding_date_as_date'] = pd.to_datetime(all_ext_feature_df['bleeding_date'], format='%d.%m.%y', errors='raise')\n",
    "uniq_donor_bleeding_date_df = all_ext_feature_df.loc[all_ext_feature_df['diagnosis_class'].isin(['cytopenia', 'MDS', 'MDS/MPN']), ['donor_id', 'bleeding_date_as_date']].drop_duplicates()\n",
    "unhealthy_with_bio_rep_donor_ids = list((uniq_donor_bleeding_date_df['donor_id'].value_counts() > 1).loc[lambda x: x].index)\n",
    "print(f'len(unhealthy_with_bio_rep_donor_ids): {len(unhealthy_with_bio_rep_donor_ids)}')\n",
    "uniq_patient_bleeding_date_df = uniq_donor_bleeding_date_df[uniq_donor_bleeding_date_df['donor_id'].isin(unhealthy_with_bio_rep_donor_ids)]\n",
    "print(uniq_patient_bleeding_date_df['donor_id'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig4'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "\n",
    "comp_class_to_num = {\n",
    "    'norm-like comp': 1,\n",
    "    'only HSC/MPP': 5,\n",
    "    'high GMP': 6,\n",
    "    'high BEMP': 4,\n",
    "    'high CLP': 2,\n",
    "    'high CLP low MEBEMP': 3,\n",
    "    'other abnormal composition': 7,\n",
    "}\n",
    "\n",
    "SHOW_COMP_CLASS = True\n",
    "SHOW_COMP_CLASS = False\n",
    "\n",
    "SHOW_DONOR_INFO = True\n",
    "SHOW_DONOR_INFO = False\n",
    "\n",
    "PLOT_LINEAR_FIT = False\n",
    "PLOT_LINEAR_FIT = True\n",
    "\n",
    "PLOT_Y_EQUALS_X = True\n",
    "PLOT_Y_EQUALS_X = False\n",
    "\n",
    "FORCE_LOWER_LIM_ZERO = True\n",
    "FORCE_LOWER_LIM_ZERO = False\n",
    "\n",
    "HORIZ_LINE_YS = [\n",
    "    # np.log2(0.05 + 0.005), # 0.05 is the BM FACS blast clinical threshold. 0.005 is the epsilon for log we used.\n",
    "]\n",
    "\n",
    "# NOTE: N413 (?) in demux_n_18_12_23_2 is discarded due to biased_composition_due_to_discarded_low_umi_count_barcodes\n",
    "    \n",
    "plt.close('all')\n",
    "curr_df = all_ext_feature_df.copy()\n",
    "\n",
    "curr_df = curr_df[curr_df['in_blast_vs_clp_e_scatter']]\n",
    "\n",
    "curr_df['c_clp_e_count'] = curr_df['c_clp_e'] * curr_df['final_c_hspc_count']\n",
    "\n",
    "curr_df = generic_utils.merge_preserving_df1_index_and_row_order(curr_df, pd.read_csv(get_mds_params()['fig4_donor_composition_class_df_csv_file_path']), how='left', allow_multiple_common_column_names=True)\n",
    "curr_df['c_clp_e_count'] = curr_df['c_clp_e'] * curr_df['final_c_hspc_count']\n",
    "\n",
    "# y_column_name = 'final_c_hspc_count'\n",
    "# y_column_name = 'c_clp_e_count'\n",
    "# x_column_name = 'c_clp_e_count'\n",
    "# x_column_name = 'c_clp_e'\n",
    "\n",
    "y_column_name = 'log_bm_FACS_blast_frac'\n",
    "x_column_name = 'log_c_clp_e'\n",
    "\n",
    "palette = clinical_data_interface_and_utils.DIAGNOSIS_CLASS_TO_COLOR.copy()\n",
    "\n",
    "x_or_y_nan_mask = curr_df[x_column_name].isna() | curr_df[y_column_name].isna()\n",
    "assert not x_or_y_nan_mask.any()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5,4))\n",
    "sb.scatterplot(\n",
    "    # data=curr_composition_and_clinical_df,\n",
    "    data=curr_df,\n",
    "    x=x_column_name,\n",
    "    y=y_column_name,\n",
    "    \n",
    "    palette=palette,\n",
    "    hue='diagnosis_class',\n",
    "    # hue='in_mds_cyto_groups_analysis',\n",
    "\n",
    "    # legend=False,\n",
    "    ax=ax,\n",
    ")\n",
    "\n",
    "if SHOW_COMP_CLASS:\n",
    "    for _, row in curr_df[curr_df['composition_class'].notna()].iterrows():\n",
    "        if (row['log_c_clp_e'] > -3) or (row['log_bm_FACS_blast_frac'] > -5):\n",
    "            ax.annotate(' ' + str(comp_class_to_num[row['composition_class']]), row[[x_column_name, y_column_name]])\n",
    "\n",
    "if SHOW_DONOR_INFO:\n",
    "    for _, row in curr_df.iterrows():\n",
    "        \n",
    "        numbered_donor_id = row['numbered_donor_id']\n",
    "        numbered_donor_id_without_tech_rep_suffix = row['numbered_donor_id_without_tech_rep_suffix']\n",
    "        \n",
    "        curr_x_and_y = curr_df.loc[curr_df['numbered_donor_id'] == numbered_donor_id, [x_column_name, y_column_name]].iloc[0].to_numpy()\n",
    "        extended_numbered_donor_id = numbered_donor_id\n",
    "        extended_numbered_donor_id = numbered_donor_id_without_tech_rep_suffix\n",
    "        # extended_numbered_donor_id += '*'\n",
    "        ax.annotate(extended_numbered_donor_id, curr_x_and_y)\n",
    "    \n",
    "    # generic_utils.plot_y_equals_x_line_on_ax(ax)\n",
    "\n",
    "\n",
    "if FORCE_LOWER_LIM_ZERO:\n",
    "    x_lims = ax.get_xlim()\n",
    "    y_lims = ax.get_ylim()\n",
    "    if (curr_df[x_column_name] >= 0).all():\n",
    "        ax.set_xlim(0, x_lims[1])\n",
    "    if (curr_df[y_column_name] >= 0).all():\n",
    "        ax.set_ylim(0, y_lims[1])\n",
    "\n",
    "if PLOT_Y_EQUALS_X:\n",
    "    generic_utils.plot_y_equals_x_line_on_ax(ax=ax)\n",
    "\n",
    "if PLOT_LINEAR_FIT:\n",
    "    slope, intercept, r_value, p_value, std_err_of_estimated_gradient = scipy.stats.linregress(\n",
    "        curr_df[x_column_name], curr_df[y_column_name],\n",
    "        # alternative='two-sided',\n",
    "    )\n",
    "    generic_utils.plot_line_on_ax(ax, slope=slope, offset=intercept, color='grey', alpha=0.5, linestyle='--')\n",
    "    linear_fit_repr = f'r={r_value:.2f}, pval={p_value:.0E}'\n",
    "    print(f'slope={slope}, intercept={intercept}')\n",
    "\n",
    "sb.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "for y in HORIZ_LINE_YS:\n",
    "    ax.axhline(y, color='red', alpha=0.2)\n",
    "    \n",
    "ax_title = f'n={len(curr_df)}'\n",
    "if PLOT_LINEAR_FIT:\n",
    "    ax_title += f', {linear_fit_repr}'\n",
    "ax.set_title(ax_title)\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.savefig(os.path.join(out_dir_path, 'fig_4H_blast_vs_clp_e.png'), dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/__sup_tables'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "mut_df = arch_mutation_interface_and_utils.get_minimal_all_mutation_df()\n",
    "mut_df['nimrod_compatible_donor_id'] = mut_df['donor_id'].replace(get_sc_rna_seq_preprocessing_params()['240723_donor_id_i_use_to_nimrod_donor_id'])\n",
    "# raise\n",
    "assert (all_ext_feature_df['exp_date'] == all_ext_feature_df['bleeding_date']).all()\n",
    "mut_df['bleeding_date'] = mut_df['exp_date']\n",
    "mut_df = generic_utils.merge_preserving_df1_index_and_row_order(\n",
    "    mut_df, pd.read_csv(get_sc_rna_seq_preprocessing_params()['blood_sample_id_df_csv_file_path']), how='left', on=['donor_id', 'bleeding_date'])\n",
    "mut_df = mut_df[mut_df['blood_sample_id'].notna()]\n",
    "mut_df['blood_sample_id'] = mut_df['blood_sample_id'].astype(int)\n",
    "mut_df['CHR'] = mut_df['CHR'].astype(str)\n",
    "mut_df['POS'] = mut_df['POS'].astype(int)\n",
    "\n",
    "mut_df = mut_df[[\n",
    "    'blood_sample_id',\n",
    "    'nimrod_compatible_donor_id',\n",
    "    'gene', \n",
    "    'CHR', 'POS', 'REF', 'ALT', \n",
    "    'mean_VAF', \n",
    "]].drop_duplicates().sort_values('blood_sample_id')\n",
    "\n",
    "print(len(mut_df))\n",
    "mut_df = mut_df[~mut_df['blood_sample_id'].isin(mds_analysis.get_unreliable_arch_blood_sample_ids(all_ext_feature_df))]\n",
    "print(len(mut_df))\n",
    "mut_df.rename(columns=get_mds_params()['col_name_to_name_in_paper']).to_csv(os.path.join(out_dir_path, 's11_fig4_CH_mutations.csv'), index=False)\n",
    "\n",
    "manually_copied_ones_df = mut_df[mut_df[['CHR', 'POS', 'mean_VAF']].duplicated(keep=False)].sort_values('nimrod_compatible_donor_id')\n",
    "if not manually_copied_ones_df.empty:\n",
    "    print(f'\\nmanually copied ones, i guess:\\n{manually_copied_ones_df}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/__sup_tables'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "curr_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "curr_df = curr_df[curr_df['cna_count'] >= 1]\n",
    "\n",
    "donor_with_cnas_ids = set(curr_df['donor_id'])\n",
    "\n",
    "curr_df = curr_df[['nimrod_compatible_donor_id', 'donor_id', 'exp_name', 'blood_sample_id']].drop_duplicates()\n",
    "\n",
    "cna_df = pd.read_csv(get_mds_params()['all_cna_attrs_df_csv_file_path'])\n",
    "cna_df = cna_df[cna_df['donor_general_manual_comment'].isna()][['donor_id', 'exp_name', 'chrom', 'whole_chrom', 'first_gene_end_pos_in_chrom', 'last_gene_end_pos_in_chrom', 'is_dup', 'is_del', 'manual_comment_for_blood_aging_paper']]\n",
    "non_exp_specific_cna_df = cna_df[cna_df['exp_name'] == 'all'].drop(columns='exp_name')\n",
    "non_exp_specific_cna_donor_ids = set(non_exp_specific_cna_df['donor_id'])\n",
    "\n",
    "exp_specific_cna_df = cna_df[cna_df['exp_name'] != 'all']\n",
    "exp_specific_cna_donor_ids = set(exp_specific_cna_df['donor_id'])\n",
    "\n",
    "assert not (non_exp_specific_cna_donor_ids & exp_specific_cna_donor_ids)\n",
    "assert (non_exp_specific_cna_donor_ids | exp_specific_cna_donor_ids) >= set(curr_df['donor_id'])\n",
    "\n",
    "curr_df = pd.concat([\n",
    "    curr_df[curr_df['donor_id'].isin(list(non_exp_specific_cna_donor_ids))].merge(non_exp_specific_cna_df),\n",
    "    curr_df[curr_df['donor_id'].isin(list(exp_specific_cna_donor_ids))].merge(exp_specific_cna_df),\n",
    "], ignore_index=True)\n",
    "\n",
    "assert set(curr_df['donor_id']) == donor_with_cnas_ids\n",
    "\n",
    "assert (curr_df['is_dup'] | curr_df['is_del']).all()\n",
    "\n",
    "curr_df['CNA type'] = curr_df['is_dup'].replace({True: 'duplication', False: 'deletion'})\n",
    "curr_df.loc[curr_df['whole_chrom'], 'CNA type'] = 'whole chromosome ' + curr_df.loc[curr_df['whole_chrom'], 'CNA type']\n",
    "\n",
    "curr_df.loc[curr_df['whole_chrom'], ['first_gene_end_pos_in_chrom', 'last_gene_end_pos_in_chrom']] = np.nan\n",
    "\n",
    "curr_df = curr_df[[\n",
    "    'blood_sample_id',\n",
    "    'nimrod_compatible_donor_id',\n",
    "    \n",
    "    'chrom',\n",
    "    # 'whole_chrom',\n",
    "    'first_gene_end_pos_in_chrom',\n",
    "    'last_gene_end_pos_in_chrom',\n",
    "    'CNA type',\n",
    "    'manual_comment_for_blood_aging_paper',\n",
    "]].drop_duplicates().sort_values(['blood_sample_id', 'chrom', 'first_gene_end_pos_in_chrom'])\n",
    "curr_df.rename(columns=get_mds_params()['col_name_to_name_in_paper']).to_csv(os.path.join(out_dir_path, 's12_fig4_CNAs.csv'), index=False)\n",
    "\n",
    "assert len(curr_df.drop(columns=['blood_sample_id', 'first_gene_end_pos_in_chrom', 'last_gene_end_pos_in_chrom']).drop_duplicates()) == 38 # # 240729: that's the number in EDF 9C\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/__sup_tables'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "skip_cell_filtering_count_cols = True\n",
    "skip_cell_filtering_count_cols = False\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "\n",
    "all_ext_feature_df['treatment'].replace({x: np.nan for x in clinical_data_interface_and_utils.UNKNOWN_TREATMENT_VALUES}, inplace=True)\n",
    "\n",
    "all_ext_feature_df['BM FACS blasts (%)'] = all_ext_feature_df['bm_FACS_blast_frac'] * 100\n",
    "all_ext_feature_df['FACS_BM_date'] = pd.to_datetime(all_ext_feature_df['FACS_BM_date'])\n",
    "all_ext_feature_df['facs_bm_date_repr'] = all_ext_feature_df.apply(\n",
    "    lambda row: str(row[\"FACS_BM_date\"].year) + ('' if row['fake_FACS_BM_month'] else ('-' + str(row[\"FACS_BM_date\"].month))), axis=1)\n",
    "\n",
    "with_vaf_classifier_final_feature_names_and_types_df = pd.read_csv(get_mds_params()['with_vaf_classifier_final_feature_names_and_types_df_csv_file_path'])\n",
    "with_vaf_classifier_final_feature_names_without_cbc = with_vaf_classifier_final_feature_names_and_types_df.loc[\n",
    "    with_vaf_classifier_final_feature_names_and_types_df['type'] != 'CBC', 'name'].tolist()\n",
    "\n",
    "\n",
    "if skip_cell_filtering_count_cols:\n",
    "    count_cols = []\n",
    "    curr_df = all_ext_feature_df.copy()\n",
    "else:\n",
    "    cell_filtering_count_dfs = []\n",
    "    prev_count_cols = None\n",
    "    for dataset_name in ['illu_mds', 'ult_mds']:\n",
    "        mds_path_dict = mds_analysis_params.get_mds_path_dict(dataset_name=dataset_name)\n",
    "        if not os.path.isfile(mds_path_dict['assigned_and_no_doublet_enirhced_mc_c_ad_file_path']):\n",
    "            raise RuntimeError('you can generate the table excluding cell filtering count cols by setting skip_cell_filtering_count_cols=True')\n",
    "        count_df, count_cols = mds_analysis.write_and_get_cell_filtering_count_df(dataset_name, use_existing=True)\n",
    "        if prev_count_cols is not None:\n",
    "            assert count_cols == prev_count_cols\n",
    "        else:\n",
    "            prev_count_cols = count_cols\n",
    "        cell_filtering_count_dfs.append(count_df)\n",
    "    cell_filtering_count_df = pd.concat(cell_filtering_count_dfs, ignore_index=True)\n",
    "\n",
    "    curr_df = generic_utils.merge_preserving_df1_index_and_row_order(all_ext_feature_df, cell_filtering_count_df)\n",
    "\n",
    "curr_df['seq_platform'] = curr_df['is_ultima'].replace({True: 'Ultima', False: 'Illumina'})\n",
    "assert curr_df['seq_platform'].isin(['Ultima', 'Illumina']).all()\n",
    "\n",
    "scrna_sample_df = curr_df[[\n",
    "    'blood_sample_id',\n",
    "    'nimrod_compatible_donor_id',\n",
    "    'donor_sex',\n",
    "    'donor_age',\n",
    "    'bleeding_date_as_date',\n",
    "    'diagnosis_class',\n",
    "    'treatment',\n",
    "    'scrna_exp_id',\n",
    "    'seq_platform',\n",
    "    'in_fig4_ref_model',\n",
    "    'in_mds_cyto_groups_analysis',\n",
    "    'in_train_set',\n",
    "    'in_test_set',\n",
    "    'in_blast_vs_clp_e_scatter',\n",
    "    'used_to_fit_hspc_compos_knn',\n",
    "    'in_orig_ref_model',\n",
    "\n",
    "    'facs_bm_date_repr',\n",
    "    'BM FACS blasts (%)',\n",
    "    'cbc_date',\n",
    "    *all_ext_feature_df.columns[all_ext_feature_df.columns.isin(list(clinical_data_interface_and_utils.CBC_COL_TO_PUBLISH_COL))],\n",
    "    \n",
    "    *with_vaf_classifier_final_feature_names_without_cbc,\n",
    "\n",
    "    'final_c_hspc_count',\n",
    "    *[x for x in count_cols if x not in {'mouse_human_doublet_count', 'not_whitelist_barcode_count', 'dc_nkt_monocyte_endothel_b_lamp3_count', 'presumably_hspc_count'}],\n",
    "\n",
    "]].sort_values(['blood_sample_id', 'scrna_exp_id'])\n",
    "\n",
    "assert len(scrna_sample_df.columns) == len(set(scrna_sample_df.columns))\n",
    "\n",
    "# curr_name = c_ad_file_path_to_name[c_ad_file_path]\n",
    "\n",
    "assert set(with_vaf_classifier_final_feature_names_and_types_df['name']) <= set(scrna_sample_df.columns)\n",
    "\n",
    "scrna_sample_df.rename(columns={\n",
    "    **mds_analysis_params.get_gene_sig_score_to_name_in_paper(), \n",
    "    **get_mds_params()['col_name_to_name_in_paper'],\n",
    "    **clinical_data_interface_and_utils.CBC_COL_TO_PUBLISH_COL,\n",
    "}).to_csv(os.path.join(out_dir_path, 's9_fig4_scRNA_sample_metadata.csv'), index=False)\n",
    "\n",
    "# df.to_csv(os.path.join(out_dir_path, 'asdf.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/__sup_tables'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "sig_name_to_name_in_paper = get_mds_params()['sig_name_to_name_in_paper']\n",
    "sig_names = list(sig_name_to_name_in_paper)\n",
    "all_genes = sorted(set(itertools.chain.from_iterable(mds_analysis_params.PB_SIG_NAME_TO_GENES[x] for x in sig_names)))\n",
    "within_state_sig_df = mds_analysis_params.get_within_state_sig_df()\n",
    "within_state_sig_names = set(within_state_sig_df['sig_name'])\n",
    "\n",
    "df = pd.DataFrame({'gene': all_genes})\n",
    "\n",
    "for sig_name in sig_names:\n",
    "    sig_genes = mds_analysis_params.PB_SIG_NAME_TO_GENES[sig_name]\n",
    "    in_silico_sig = not (sig_name in within_state_sig_names)\n",
    "    col = sig_name_to_name_in_paper[sig_name]\n",
    "    if in_silico_sig:\n",
    "        col += ' (used for in-silico sorting)'\n",
    "    \n",
    "    df[col] = df['gene'].isin(sig_genes)\n",
    "\n",
    "df.rename(columns=get_mds_params()['col_name_to_name_in_paper']).to_csv(os.path.join(out_dir_path, 's13_fig4_gene_signatures.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/__sup_tables'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "unified_mc_model_name = 'final_mds_cyto_normal_excluding_atlas'\n",
    "\n",
    "all_genes = {\n",
    "    *lateral_and_noisy_genes.ULT_ILL_PROBLEMATIC_GENES,\n",
    "    *get_mds_params()['noisy_gene_names']['sex_diff_expressed'],\n",
    "}\n",
    "\n",
    "ult_ill_problematic_or_sex_diff_genes = set(lateral_and_noisy_genes.ULT_ILL_PROBLEMATIC_GENES) | set(get_mds_params()['noisy_gene_names']['sex_diff_expressed'])\n",
    "print(f'len(ult_ill_problematic_or_sex_diff_genes): {len(ult_ill_problematic_or_sex_diff_genes)}')\n",
    "\n",
    "print(f'len(lateral_and_noisy_genes.ULT_ILL_PROBLEMATIC_GENES): {len(set(lateral_and_noisy_genes.ULT_ILL_PROBLEMATIC_GENES))}')\n",
    "print(f\"len(get_mds_params()['noisy_gene_names']['sex_diff_expressed']): {len(set(get_mds_params()['noisy_gene_names']['sex_diff_expressed']))}\")\n",
    "\n",
    "col_to_genes = {}\n",
    "desc_to_illu_and_ult_genes = collections.defaultdict(set)\n",
    "for gene_desc in [\n",
    "    'genes_excluded_from_cna_due_to_consistent_obs_exp_diff',\n",
    "    'genes_excluded_from_cna_due_to_low_expr',\n",
    "    'genes_excluded_from_cna_due_to_missing_position_or_on_y',\n",
    "]:\n",
    "    for dataset_name in ('ult_mds', 'illu_mds'):\n",
    "        unified_mc_model_paths = mds_analysis_params.get_mc_model_paths(unified_mc_model_name, dataset_name)\n",
    "        curr_genes = generic_utils.read_text_file(\n",
    "            unified_mc_model_paths[gene_desc]).splitlines()\n",
    "        col_to_genes[f'{dataset_name}_{gene_desc}'] = curr_genes\n",
    "        \n",
    "        all_genes |= set(curr_genes)\n",
    "        desc_to_illu_and_ult_genes[gene_desc] |= set(curr_genes)\n",
    "desc_to_illu_and_ult_genes = dict(desc_to_illu_and_ult_genes) # I don't want a defaultdict moving around.\n",
    "for gene_desc, genes in desc_to_illu_and_ult_genes.items():\n",
    "    print(f'{gene_desc}: {len(set(genes))}')\n",
    "\n",
    "df = pd.DataFrame({'gene': sorted(all_genes)})\n",
    "df['sex-specific'] = df['gene'].isin(get_mds_params()['noisy_gene_names']['sex_diff_expressed'])\n",
    "df['sequencing platform-specific'] = df['gene'].isin(lateral_and_noisy_genes.ULT_ILL_PROBLEMATIC_GENES)\n",
    "for col, genes in col_to_genes.items():\n",
    "    df[col] = df['gene'].isin(genes)\n",
    "\n",
    "\n",
    "\n",
    "df.rename(columns={\n",
    "    **get_mds_params()['col_name_to_name_in_paper'],\n",
    "    'ult_mds_genes_excluded_from_cna_due_to_consistent_obs_exp_diff': 'consistent difference between observed and projected (Ultima)',\n",
    "    'ult_mds_genes_excluded_from_cna_due_to_low_expr': 'lowly expressed (Ultima)',\n",
    "    'ult_mds_genes_excluded_from_cna_due_to_missing_position_or_on_y': 'missing chromosomal coordinates or on Y chromosome (Ultima)',\n",
    "    'illu_mds_genes_excluded_from_cna_due_to_consistent_obs_exp_diff': 'consistent difference between observed and projected (Illumina)',\n",
    "    'illu_mds_genes_excluded_from_cna_due_to_low_expr': 'lowly expressed (Illumina)',\n",
    "    'illu_mds_genes_excluded_from_cna_due_to_missing_position_or_on_y': 'missing chromosomal coordinates or on Y chromosome (Illumina)',\n",
    "}).to_csv(os.path.join(out_dir_path, 's14_fig4_genes_excluded_from_CNA_analysis.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = 'temp/_fig_sup_4/N200'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "karyo_out_dir_path = mds_analysis_params.get_mds_path_dict(dataset_name='ult_mds')['karyotype_estimation_out_dir_path']\n",
    "\n",
    "for exp_name, N200_del_5q_thresh, N200_dup_8_thresh in [\n",
    "    ('demux_26_03_23_1', -0.2, 0.24),\n",
    "    ('demux_11_04_21_2', -0.2, 0.18),\n",
    "    ('demux_10_01_22_1', -0.25, 0.21),\n",
    "]:\n",
    "    mc_ad_file_path = mds_analysis_params.get_mc_model_paths(f'final_only_N200_only_{exp_name}', dataset_name='ult_mds')['metacells_with_projection_ad']\n",
    "    mc_ad = ad.read_h5ad(mc_ad_file_path)\n",
    "    df_csv_file_path = f'{karyo_out_dir_path}/N200/{exp_name}/corrected_by_gc/mc_hists/cna_median_normalized_projected_fold_df.csv'\n",
    "    plt.close('all')\n",
    "    df = pd.read_csv(df_csv_file_path)\n",
    "    assert len(df) == mc_ad.n_obs\n",
    "\n",
    "    hspc_mask = mc_ad.obs['state'].isin(mds_analysis_params.PB_HSPC_STATE_NAMES).to_numpy()\n",
    "    df = df[hspc_mask]\n",
    "    # print(hspc_mask.mean())\n",
    "\n",
    "    # raise\n",
    "    fig, ax = plt.subplots()\n",
    "    sb.scatterplot(df, x='N200_del_5q', y='N200_dup_8', hue='N200_del_3p_end_and_3q_start', s=15, ax=ax)\n",
    "    ax.axvline(N200_del_5q_thresh, color='red', alpha=0.3)\n",
    "    ax.axhline(N200_dup_8_thresh, color='red', alpha=0.3)\n",
    "\n",
    "    mutated_mc_frac = ((df['N200_del_5q'] < N200_del_5q_thresh) & (df['N200_dup_8'] > N200_dup_8_thresh)).mean()\n",
    "    ax.set_title(f'{exp_name}, mutated_mc_frac: {mutated_mc_frac:.3f}')\n",
    "    sb.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1))\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(os.path.join(out_dir_path, f'for_EDF_12A_{exp_name}_mutated_mc_by_del_5q_and_dup_8_scatter.png'), dpi=300)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir_path = f'temp/_fig_sup_4/CBC_history_case_studies'\n",
    "pathlib.Path(out_dir_path).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "all_ext_feature_df = pd.read_csv(get_mds_params()['all_ext_donor_feature_df_csv_file_path'])\n",
    "all_ext_feature_df['bleeding_date_as_date'] = pd.to_datetime(all_ext_feature_df['bleeding_date'], format='%d.%m.%y', errors='raise')\n",
    "cbc_df = pd.read_csv(get_sc_rna_seq_preprocessing_params()['donor_table_paths']['minimal_cbc_df_csv']).drop_duplicates()\n",
    "cbc_df['cbc_date'] = pd.to_datetime(cbc_df['cbc_date'], errors='raise')\n",
    "\n",
    "for donor_id in ('N200', 'N211'):\n",
    "    curr_df = cbc_df[cbc_df['donor_id'] == donor_id]\n",
    "    tenx_dates = all_ext_feature_df.loc[all_ext_feature_df['donor_id'] == donor_id, 'bleeding_date_as_date']\n",
    "\n",
    "    for y_col in (\n",
    "        'Hemoglobin (g/dl)', \n",
    "        # 'RDW (%)',\n",
    "    ):\n",
    "        plt.close('all')\n",
    "        fig, ax = plt.subplots(figsize=(15,3))\n",
    "        sb.scatterplot(curr_df, x='cbc_date', y=y_col, ax=ax)\n",
    "        # ax.set_xticklabels(\n",
    "        #     ax.get_xticklabels(),\n",
    "        #     rotation=45, ha='right', \n",
    "        #     rotation_mode='anchor', \n",
    "        #     # fontsize='small',\n",
    "        # )\n",
    "        ax.set_title(donor_id)\n",
    "        ax.set_xlabel(None)\n",
    "        for x in tenx_dates:\n",
    "            print(x)\n",
    "            ax.axvline(x, color='grey', alpha=0.5)\n",
    "        fig.tight_layout()\n",
    "        fig.savefig(os.path.join(out_dir_path, f'EDF_12AB_{donor_id}_{generic_utils.strip_special_chars_on_edges_and_replace_others_with_underscores(y_col)}_history.png'), dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "vscode": {
   "interpreter": {
    "hash": "377104351be6dc5fb44deacee428a5d5a5aaea0f9f6aa21ce74220904ae67020"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
